From 01b16b2a9aacb70922043f2f1e0944adfaf3ce89 Mon Sep 17 00:00:00 2001
From: Maharaja Kennadyrajan <quic_mkenna@quicinc.com>
Date: Thu, 8 Feb 2024 12:49:21 +0530
Subject: [PATCH] wifi: ath12k: Enhance the htt stats

Enhance the below htt stats types with new members added
in the stats struct.

Stats type added/updated are 1, 4, 9, 10, 12 and 22.

Also fixed the display issue in the stats 9 and 10.

Signed-off-by: Maharaja Kennadyrajan <quic_mkenna@quicinc.com>
---
 .../wireless/ath/ath12k/debugfs_htt_stats.c   | 235 +++++++++++++++---
 .../wireless/ath/ath12k/debugfs_htt_stats.h   | 111 ++++++++-
 2 files changed, 310 insertions(+), 36 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
@@ -1301,6 +1301,7 @@ static inline void htt_print_tx_sounding
 						   struct debug_htt_stats_req *stats_req)
 {
 	s32 i;
+	s32 tag_len = 0;
 	const struct htt_tx_sounding_stats_tlv *htt_stats_buf = tag_buf;
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
@@ -1309,6 +1310,7 @@ static inline void htt_print_tx_sounding
 	const u32 *cbf_40 = htt_stats_buf->cbf_40;
 	const u32 *cbf_80 = htt_stats_buf->cbf_80;
 	const u32 *cbf_160 = htt_stats_buf->cbf_160;
+	const u32 *cbf_320 = htt_stats_buf->cbf_320;
 
 	if (htt_stats_buf->tx_sounding_mode == HTT_TX_AC_SOUNDING_MODE) {
 		len += scnprintf(buf + len, buf_len - len,
@@ -1392,6 +1394,219 @@ static inline void htt_print_tx_sounding
 					 htt_stats_buf->sounding[2],
 					 htt_stats_buf->sounding[3]);
 		}
+	} else if (htt_stats_buf->tx_sounding_mode == HTT_TX_BE_SOUNDING_MODE) {
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nHTT_TX_BE_SOUNDING_STATS_TLV:\n");
+		len += scnprintf(buf + len, buf_len - len,
+				 "be_cbf_20 = IBF : %u, SU_SIFS : %u, SU_RBO : %u, MU_SIFS : %u, MU_RBO : %u\n",
+				 cbf_20[HTT_IMPLICIT_TXBF_STEER_STATS],
+				 cbf_20[HTT_EXPLICIT_TXBF_SU_SIFS_STEER_STATS],
+				 cbf_20[HTT_EXPLICIT_TXBF_SU_RBO_STEER_STATS],
+				 cbf_20[HTT_EXPLICIT_TXBF_MU_SIFS_STEER_STATS],
+				 cbf_20[HTT_EXPLICIT_TXBF_MU_RBO_STEER_STATS]);
+		len += scnprintf(buf + len, buf_len - len,
+				 "be_cbf_40 = IBF : %u, SU_SIFS : %u, SU_RBO : %u, MU_SIFS : %u, MU_RBO : %u\n",
+				 cbf_40[HTT_IMPLICIT_TXBF_STEER_STATS],
+				 cbf_40[HTT_EXPLICIT_TXBF_SU_SIFS_STEER_STATS],
+				 cbf_40[HTT_EXPLICIT_TXBF_SU_RBO_STEER_STATS],
+				 cbf_40[HTT_EXPLICIT_TXBF_MU_SIFS_STEER_STATS],
+				 cbf_40[HTT_EXPLICIT_TXBF_MU_RBO_STEER_STATS]);
+		len += scnprintf(buf + len, buf_len - len,
+				 "be_cbf_80 = IBF : %u, SU_SIFS : %u, SU_RBO : %u, MU_SIFS : %u, MU_RBO : %u\n",
+				 cbf_80[HTT_IMPLICIT_TXBF_STEER_STATS],
+				 cbf_80[HTT_EXPLICIT_TXBF_SU_SIFS_STEER_STATS],
+				 cbf_80[HTT_EXPLICIT_TXBF_SU_RBO_STEER_STATS],
+				 cbf_80[HTT_EXPLICIT_TXBF_MU_SIFS_STEER_STATS],
+				 cbf_80[HTT_EXPLICIT_TXBF_MU_RBO_STEER_STATS]);
+		len += scnprintf(buf + len, buf_len - len,
+				 "be_cbf_160 = IBF : %u, SU_SIFS : %u, SU_RBO : %u, MU_SIFS : %u, MU_RBO : %u\n",
+				 cbf_160[HTT_IMPLICIT_TXBF_STEER_STATS],
+				 cbf_160[HTT_EXPLICIT_TXBF_SU_SIFS_STEER_STATS],
+				 cbf_160[HTT_EXPLICIT_TXBF_SU_RBO_STEER_STATS],
+				 cbf_160[HTT_EXPLICIT_TXBF_MU_SIFS_STEER_STATS],
+				 cbf_160[HTT_EXPLICIT_TXBF_MU_RBO_STEER_STATS]);
+		len += scnprintf(buf + len, buf_len - len,
+				 "be_cbf_320 = IBF : %u, SU_SIFS : %u, SU_RBO : %u, MU_SIFS : %u, MU_RBO : %u\n",
+				 cbf_320[HTT_IMPLICIT_TXBF_STEER_STATS],
+				 cbf_320[HTT_EXPLICIT_TXBF_SU_SIFS_STEER_STATS],
+				 cbf_320[HTT_EXPLICIT_TXBF_SU_RBO_STEER_STATS],
+				 cbf_320[HTT_EXPLICIT_TXBF_MU_SIFS_STEER_STATS],
+				 cbf_320[HTT_EXPLICIT_TXBF_MU_RBO_STEER_STATS]);
+
+		for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS; i++) {
+			len += scnprintf(buf + len, buf_len - len,
+					 "Sounding User %u = 20MHz: %u, 40MHz : %u, 80MHz: %u, 160MHz: %u, 320MHz: %u\n",
+					 i,
+					 htt_stats_buf->sounding[0],
+					 htt_stats_buf->sounding[1],
+					 htt_stats_buf->sounding[2],
+					 htt_stats_buf->sounding[3],
+					 htt_stats_buf->sounding_320[i]);
+		}
+	} else if (htt_stats_buf->tx_sounding_mode == HTT_TX_CMN_SOUNDING_MODE) {
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nCV UPLOAD HANDLER STATS:-\n");
+		len += scnprintf(buf + len, buf_len - len, "cv_nc_mismatch_err : %u\n",
+				 htt_stats_buf->cv_nc_mismatch_err);
+		len += scnprintf(buf + len, buf_len - len, "cv_fcs_err : %u\n",
+				 htt_stats_buf->cv_fcs_err);
+		len += scnprintf(buf + len, buf_len - len, "cv_frag_idx_mismatch : %u\n",
+				 htt_stats_buf->cv_frag_idx_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_invalid_peer_id : %u\n",
+				 htt_stats_buf->cv_invalid_peer_id);
+		len += scnprintf(buf + len, buf_len - len, "cv_no_txbf_setup : %u\n",
+				 htt_stats_buf->cv_no_txbf_setup);
+		len += scnprintf(buf + len, buf_len - len, "cv_expiry_in_update : %u\n",
+				 htt_stats_buf->cv_expiry_in_update);
+		len += scnprintf(buf + len, buf_len - len, "cv_pkt_bw_exceed : %u\n",
+				 htt_stats_buf->cv_pkt_bw_exceed);
+		len += scnprintf(buf + len, buf_len - len, "cv_dma_not_done_err : %u\n",
+				 htt_stats_buf->cv_dma_not_done_err);
+		len += scnprintf(buf + len, buf_len - len, "cv_update_failed : %u\n",
+				 htt_stats_buf->cv_update_failed);
+		len += scnprintf(buf + len, buf_len - len, "cv_dma_timeout_error : %u\n",
+				 htt_stats_buf->cv_dma_timeout_error);
+		len += scnprintf(buf + len, buf_len - len, "cv_buf_ibf_uploads : %u\n",
+				 htt_stats_buf->cv_buf_ibf_uploads);
+		len += scnprintf(buf + len, buf_len - len, "cv_buf_ebf_uploads : %u\n",
+				 htt_stats_buf->cv_buf_ebf_uploads);
+		len += scnprintf(buf + len, buf_len - len, "cv_buf_received : %u\n",
+				 htt_stats_buf->cv_buf_received);
+		len += scnprintf(buf + len, buf_len - len, "cv_buf_fed_back : %u\n",
+				 htt_stats_buf->cv_buf_fed_back);
+		len += scnprintf(buf + len, buf_len - len, "\nCV QUERY STATS:-\n");
+		len += scnprintf(buf + len, buf_len - len, "cv_total_query : %u\n",
+				 htt_stats_buf->cv_total_query);
+		len += scnprintf(buf + len, buf_len - len, "cv_total_pattern_query : %u\n",
+				 htt_stats_buf->cv_total_pattern_query);
+		len += scnprintf(buf + len, buf_len - len, "cv_total_bw_query : %u\n",
+				 htt_stats_buf->cv_total_bw_query);
+		len += scnprintf(buf + len, buf_len - len, "cv_invalid_bw_coding : %u\n",
+				 htt_stats_buf->cv_invalid_bw_coding);
+		len += scnprintf(buf + len, buf_len - len, "cv_forced_sounding : %u\n",
+				 htt_stats_buf->cv_forced_sounding);
+		len += scnprintf(buf + len, buf_len - len, "cv_standalone_sounding : %u\n",
+				 htt_stats_buf->cv_standalone_sounding);
+		len += scnprintf(buf + len, buf_len - len, "cv_nc_mismatch : %u\n",
+				 htt_stats_buf->cv_nc_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_fb_type_mismatch : %u\n",
+				 htt_stats_buf->cv_fb_type_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_ofdma_bw_mismatch : %u\n",
+				 htt_stats_buf->cv_ofdma_bw_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_bw_mismatch : %u\n",
+				 htt_stats_buf->cv_bw_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_pattern_mismatch : %u\n",
+				 htt_stats_buf->cv_pattern_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_preamble_mismatch : %u\n",
+				 htt_stats_buf->cv_preamble_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_nr_mismatch : %u\n",
+				 htt_stats_buf->cv_nr_mismatch);
+		len += scnprintf(buf + len, buf_len - len, "cv_in_use_cnt_exceeded : %u\n",
+				 htt_stats_buf->cv_in_use_cnt_exceeded);
+		len += scnprintf(buf + len, buf_len - len, "cv_ntbr_sounding : %u\n",
+				 htt_stats_buf->cv_ntbr_sounding);
+		len += scnprintf(buf + len, buf_len - len, "cv_found_upload_in_progress : %u\n",
+				 htt_stats_buf->cv_found_upload_in_progress);
+		len += scnprintf(buf + len, buf_len - len, "cv_expired_during_query : %u\n",
+				 htt_stats_buf->cv_expired_during_query);
+		len += scnprintf(buf + len, buf_len - len, "cv_found : %u\n",
+				 htt_stats_buf->cv_found);
+		len += scnprintf(buf + len, buf_len - len, "cv_not_found : %u\n",
+				 htt_stats_buf->cv_not_found);
+		len += scnprintf(buf + len, buf_len - len, "cv_total_query_ibf : %u\n",
+				 htt_stats_buf->cv_total_query_ibf);
+		len += scnprintf(buf + len, buf_len - len, "cv_found_ibf : %u\n",
+				 htt_stats_buf->cv_found_ibf);
+		len += scnprintf(buf + len, buf_len - len, "cv_not_found_ibf : %u\n",
+				 htt_stats_buf->cv_not_found_ibf);
+		len += scnprintf(buf + len, buf_len - len, "cv_expired_during_query_ibf : %u\n",
+				 htt_stats_buf->cv_expired_during_query_ibf);
+		len += scnprintf(buf + len, buf_len - len, "adaptive_snd_total_query : %u\n",
+				 htt_stats_buf->adaptive_snd_total_query);
+
+		tag_len = (HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS +
+				HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS);
+
+		len += scnprintf(buf + len, buf_len - len, "adaptive_snd_total_mcs_drop: ");
+		for (i = 0; i < tag_len; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u:%u ",
+					 i, htt_stats_buf->adaptive_snd_total_mcs_drop[i]);
+
+		len += scnprintf(buf + len, buf_len - len, "\nadaptive_snd_kicked_in : %u\n",
+				 htt_stats_buf->adaptive_snd_kicked_in);
+		len += scnprintf(buf + len, buf_len - len, "adaptive_snd_back_to_default : %u\n",
+				 htt_stats_buf->adaptive_snd_back_to_default);
+
+	} else if (htt_stats_buf->tx_sounding_mode == HTT_TX_CV_CORR_MODE) {
+		len += scnprintf(buf + len, buf_len - len, "\nCV CORRELATION TRIGGER STATS:-\n");
+		len += scnprintf(buf + len, buf_len - len, "cv_corr_trigger_online_mode : %u\n",
+				 htt_stats_buf->cv_corr_trigger_online_mode);
+		len += scnprintf(buf + len, buf_len - len, "cv_corr_trigger_offline_mode : %u\n",
+				 htt_stats_buf->cv_corr_trigger_offline_mode);
+		len += scnprintf(buf + len, buf_len - len, "cv_corr_trigger_hybrid_mode : %u\n",
+				 htt_stats_buf->cv_corr_trigger_hybrid_mode);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_trigger_computation_level_0 : %u\n",
+				 htt_stats_buf->cv_corr_trigger_computation_level_0);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_trigger_computation_level_1 : %u\n",
+				 htt_stats_buf->cv_corr_trigger_computation_level_1);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_trigger_computation_level_2 : %u\n",
+				 htt_stats_buf->cv_corr_trigger_computation_level_2);
+
+		len += scnprintf(buf + len, buf_len - len, "cv_corr_trigger_num_users : ");
+		for (i = 0; i < HTT_TX_CV_CORR_MAX_NUM_COLUMNS; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u_user: %u ",
+					 (i + 1), htt_stats_buf->cv_corr_trigger_num_users[i]);
+
+		len += scnprintf(buf + len, buf_len - len, "\ncv_corr_trigger_num_streams : ");
+		for (i = 0; i < HTT_TX_CV_CORR_MAX_NUM_COLUMNS; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u_stream: %u ",
+					 (i + 1), htt_stats_buf->cv_corr_trigger_num_streams[i]);
+
+		len += scnprintf(buf + len, buf_len - len, "\n\nCV CORRELATION UPLOAD STATS:-\n");
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_total_buf_received : %u\n",
+				 htt_stats_buf->cv_corr_upload_total_buf_received);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_total_buf_fed_back : %u\n",
+				 htt_stats_buf->cv_corr_upload_total_buf_fed_back);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_total_processing_failed : %u\n",
+				 htt_stats_buf->cv_corr_upload_total_processing_failed);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_failed_total_users_zero : %u\n",
+				 htt_stats_buf->cv_corr_upload_failed_total_users_zero);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_failed_total_users_exceeded : %u\n",
+				 htt_stats_buf->cv_corr_upload_failed_total_users_exceeded);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_failed_peer_not_found : %u\n",
+				 htt_stats_buf->cv_corr_upload_failed_peer_not_found);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_user_nss_exceeded : %u\n",
+				 htt_stats_buf->cv_corr_upload_user_nss_exceeded);
+		len += scnprintf(buf + len, buf_len - len,
+				 "cv_corr_upload_invalid_lookup_index : %u\n",
+				 htt_stats_buf->cv_corr_upload_invalid_lookup_index);
+
+		len += scnprintf(buf + len, buf_len - len, "cv_corr_upload_total_num_users : ");
+		for (i = 0; i < HTT_TX_CV_CORR_MAX_NUM_COLUMNS; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u_user: %u ", (i + 1),
+					 htt_stats_buf->cv_corr_upload_total_num_users[i]);
+
+		len += scnprintf(buf + len, buf_len - len, "\ncv_corr_upload_total_num_streams : ");
+		for (i = 0; i < HTT_TX_CV_CORR_MAX_NUM_COLUMNS; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u_stream: %u ", (i + 1),
+					 htt_stats_buf->cv_corr_upload_total_num_streams[i]);
+
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
 	stats_req->buf_len = len;
@@ -1446,6 +1661,9 @@ htt_print_tx_selfgen_cmn_stats_tlv(const
 			 htt_stats_buf->su_sw_rts_flushed);
 	len += scnprintf(buf + len, buf_len - len, "su_sw_rts_rcvd_cts_diff_bw = %u\n",
 			 htt_stats_buf->su_sw_rts_rcvd_cts_diff_bw);
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->smart_basic_trig_sch_histogram,
+			   "smart_basic_trig_sch_histogram",
+			   HTT_MAX_NUM_SBT_INTR, "\n");
 
 	stats_req->buf_len = len;
 }
@@ -1462,11 +1680,11 @@ htt_print_tx_selfgen_ac_stats_tlv(const
 	len += scnprintf(buf + len, buf_len - len, "HTT_TX_SELFGEN_AC_STATS_TLV:\n");
 	len += scnprintf(buf + len, buf_len - len, "ac_su_ndpa_queued = %u\n",
 			   htt_stats_buf->ac_su_ndpa_queued);
-	len += scnprintf(buf + len, buf_len - len, "ac_su_ndpa_tried = %u",
+	len += scnprintf(buf + len, buf_len - len, "ac_su_ndpa_tried = %u\n",
 			   htt_stats_buf->ac_su_ndpa);
 	len += scnprintf(buf + len, buf_len - len, "ac_su_ndp_queued = %u\n",
 			   htt_stats_buf->ac_su_ndp_queued);
-	len += scnprintf(buf + len, buf_len - len, "ac_su_ndp_tried = %u",
+	len += scnprintf(buf + len, buf_len - len, "ac_su_ndp_tried = %u\n",
 			   htt_stats_buf->ac_su_ndp);
 	len += scnprintf(buf + len, buf_len - len, "ac_mu_mimo_ndpa_queued = %u\n",
 			   htt_stats_buf->ac_mu_mimo_ndpa_queued);
@@ -1474,7 +1692,7 @@ htt_print_tx_selfgen_ac_stats_tlv(const
 			   htt_stats_buf->ac_mu_mimo_ndpa);
 	len += scnprintf(buf + len, buf_len - len, "ac_mu_mimo_ndp_queued = %u\n",
 			   htt_stats_buf->ac_mu_mimo_ndp_queued);
-	len += scnprintf(buf + len, buf_len - len, "ac_mu_mimo_ndp_tried = %u",
+	len += scnprintf(buf + len, buf_len - len, "ac_mu_mimo_ndp_tried = %u\n",
 			   htt_stats_buf->ac_mu_mimo_ndp);
 	len += scnprintf(buf + len, buf_len - len, "ac_mu_mimo_brpoll_1_queued = %u\n",
 			   htt_stats_buf->ac_mu_mimo_brpoll_1_queued);
@@ -1531,12 +1749,50 @@ htt_print_tx_selfgen_ax_stats_tlv(const
 			   HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS, "\n");
 	len += scnprintf(buf + len, buf_len - len, "ax_basic_trigger = %u\n",
 			 htt_stats_buf->ax_basic_trigger);
+	len += scnprintf(buf + len, buf_len - len, "ax_ulmumimo_total_trigger = %u\n",
+			 htt_stats_buf->ax_ulmumimo_trigger);
 	len += scnprintf(buf + len, buf_len - len, "ax_bsr_trigger = %u\n",
 			 htt_stats_buf->ax_bsr_trigger);
 	len += scnprintf(buf + len, buf_len - len, "ax_mu_bar_trigger = %u\n",
 			 htt_stats_buf->ax_mu_bar_trigger);
 	len += scnprintf(buf + len, buf_len - len, "ax_mu_rts_trigger = %u\n",
 			 htt_stats_buf->ax_mu_rts_trigger);
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->combined_ax_bsr_trigger_tried,
+			   "combined_ax_bsr_trigger_tried",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->combined_ax_bsr_trigger_err,
+			   "combined_ax_bsr_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->standalone_ax_bsr_trigger_tried,
+			   "standalone_ax_bsr_trigger_tried",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->standalone_ax_bsr_trigger_err,
+			   "standalone_ax_bsr_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_ax_su_ulofdma_basic_trigger,
+			   "manual_ax_su_ulofdma_basic_trigger",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_ax_su_ulofdma_basic_trigger_err,
+			   "manual_ax_su_ulofdma_basic_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_ax_mu_ulofdma_basic_trigger,
+			   "manual_ax_mu_ulofdma_basic_trigger",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_ax_mu_ulofdma_basic_trigger_err,
+			   "manual_ax_mu_ulofdma_basic_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_basic_trigger_per_ac,
+			   "ax_basic_trigger_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_basic_trigger_errors_per_ac,
+			   "ax_basic_trigger_errors_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_bar_trigger_per_ac,
+			   "ax_mu_bar_trigger_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_bar_trigger_errors_per_ac,
+			   "ax_mu_bar_trigger_errors_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
 
 	stats_req->buf_len = len;
 }
@@ -1634,6 +1890,12 @@ htt_print_tx_selfgen_ax_err_stats_tlv(co
 			 htt_stats_buf->ax_mu_bar_trigger_err);
 	len += scnprintf(buf + len, buf_len - len, "ax_mu_rts_trigger_err = %u\n",
 			 htt_stats_buf->ax_mu_rts_trigger_err);
+	len += scnprintf(buf + len, buf_len - len, "ax_basic_trigger_partial_resp = %u\n",
+			 htt_stats_buf->ax_basic_trigger_partial_resp);
+	len += scnprintf(buf + len, buf_len - len, "ax_bsr_trigger_partial_resp = %u\n",
+			 htt_stats_buf->ax_bsr_trigger_partial_resp);
+	len += scnprintf(buf + len, buf_len - len, "ax_mu_bar_trigger_partial_resp = %u\n",
+			 htt_stats_buf->ax_mu_bar_trigger_partial_resp);
 
 	stats_req->buf_len = len;
 }
@@ -1648,31 +1910,31 @@ void htt_print_tx_selfgen_ac_sched_statu
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
 
 	len += scnprintf(buf + len, buf_len - len,
-			 "HTT_TX_SELFGEN_AC_SCHED_STATUS_STATS_TLV:");
+			 "HTT_TX_SELFGEN_AC_SCHED_STATUS_STATS_TLV:\n");
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_su_ndpa_sch_status,
-			   "ac_su_ndpa_sch_status = %s ",
+			   "ac_su_ndpa_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_su_ndp_sch_status,
-			   "ac_su_ndp_sch_status = %s ",
+			   "ac_su_ndp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_ndpa_sch_status,
-			   "ac_mu_mimo_ndpa_sch_status = %s ",
+			   "ac_mu_mimo_ndpa_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_ndp_sch_status,
-			   "ac_mu_mimo_ndp_sch_status = %s ",
+			   "ac_mu_mimo_ndp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_brp_sch_status,
-			   "ac_mu_mimo_brp_sch_status = %s ",
+			   "ac_mu_mimo_brp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_su_ndp_sch_flag_err,
-			   "ac_su_ndp_sch_flag_err = %s ",
+			   "ac_su_ndp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_ndp_sch_flag_err,
-			   "ac_mu_mimo_ndp_sch_flag_err = %s ",
+			   "ac_mu_mimo_ndp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_brp_sch_flag_err,
-			   "ac_mu_mimo_brp_sch_flag_err = %s ",
+			   "ac_mu_mimo_brp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 
 	stats_req->buf_len = len;
@@ -1688,43 +1950,43 @@ htt_print_tx_selfgen_ax_sched_status_sta
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
 
 	len += scnprintf(buf + len, buf_len - len,
-			 "HTT_TX_SELFGEN_AX_SCHED_STATUS_STATS_TLV:");
+			 "HTT_TX_SELFGEN_AX_SCHED_STATUS_STATS_TLV:\n");
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_su_ndpa_sch_status,
-			   "ax_su_ndpa_sch_status = %s ",
+			   "ax_su_ndpa_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_su_ndp_sch_status,
-			   "ax_su_ndp_sch_status = %s ",
+			   "ax_su_ndp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_ndpa_sch_status,
-			   "ax_mu_mimo_ndpa_sch_status = %s ",
+			   "ax_mu_mimo_ndpa_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_ndp_sch_status,
-			   "ax_mu_mimo_ndp_sch_status = %s ",
+			   "ax_mu_mimo_ndp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_brp_sch_status,
-			   "ax_mu_brp_sch_status = %s ",
+			   "ax_mu_brp_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_bar_sch_status,
-			   "ax_mu_bar_sch_status = %s ",
+			   "ax_mu_bar_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_basic_trig_sch_status,
-			   "ax_basic_trig_sch_status = %s ",
+			   "ax_basic_trig_sch_status",
 			   HTT_TX_PDEV_STATS_NUM_TX_ERR_STATUS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_su_ndp_sch_flag_err,
-			   "ax_su_ndp_sch_flag_err = %s ",
+			   "ax_su_ndp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_ndp_sch_flag_err,
-			   "ax_mu_mimo_ndp_sch_flag_err = %s ",
+			   "ax_mu_mimo_ndp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_brp_sch_flag_err,
-			   "ax_mu_brp_sch_flag_err = %s ",
+			   "ax_mu_brp_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_bar_sch_flag_err,
-			   "ax_mu_bar_sch_flag_err = %s ",
+			   "ax_mu_bar_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_basic_trig_sch_flag_err,
-			   "ax_basic_trig_sch_flag_err = %s ",
+			   "ax_basic_trig_sch_flag_err",
 			   HTT_TX_SELFGEN_NUM_SCH_TSFLAG_ERROR_STATS, "\n");
 
 	stats_req->buf_len = len;
@@ -2122,7 +2384,7 @@ htt_print_sched_txq_cmd_posted_tlv_v(con
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
-	u16 num_elements = min_t(u16, (tag_len >> 2), HTT_TX_PDEV_SCHED_TX_MODE_MAX);
+	u16 num_elements = tag_len >> 2;
 
 	len += scnprintf(buf + len, buf_len - len, "HTT_SCHED_TXQ_CMD_POSTED_TLV_V:\n");
 
@@ -2141,7 +2403,7 @@ htt_print_sched_txq_cmd_reaped_tlv_v(con
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
-	u16 num_elements = min_t(u16, (tag_len >> 2), HTT_TX_PDEV_SCHED_TX_MODE_MAX);
+	u16 num_elements = tag_len >> 2;
 
 	len += scnprintf(buf + len, buf_len - len, "HTT_SCHED_TXQ_CMD_REAPED_TLV_V:\n");
 
@@ -3201,29 +3463,32 @@ static inline void htt_print_tx_pdev_rat
 			 htt_stats_buf->tx_legacy_ofdm_rate[6],
 			 htt_stats_buf->tx_legacy_ofdm_rate[7]);
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_mcs, "tx_mcs",
-			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_mcs_ext, "tx_mcs_ext",
-			   HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+	len += scnprintf(buf + len, buf_len - len, "tx_mcs -2: %u, -1: %u, ",
+			 htt_stats_buf->tx_mcs_ext_2[0], htt_stats_buf->tx_mcs_ext_2[1]);
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_mcs, " ",
+			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->tx_mcs_ext[0], htt_stats_buf->tx_mcs_ext[1]);
+
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_tx_mcs,
-			   "ac_mu_mimo_tx_mcs", HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_mcs_ext,
-			   "ac_mu_mimo_tx_mcs_ext",
-			   HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+			   "ac_mu_mimo_tx_mcs", HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->ax_mu_mimo_tx_mcs_ext[0],
+			 htt_stats_buf->ax_mu_mimo_tx_mcs_ext[1]);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_mcs,
 			   "ax_mu_mimo_tx_mcs", HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_mcs_ext,
-			   "ax_mu_mimo_tx_mcs_ext",
-			   HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->ax_mu_mimo_tx_mcs_ext[0],
+			 htt_stats_buf->ax_mu_mimo_tx_mcs_ext[1]);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_mcs, "ofdma_tx_mcs",
-			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
+			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_mcs_ext,
-			   "ofdma_tx_mcs_ext",
-			   HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->ofdma_tx_mcs_ext[0],
+			 htt_stats_buf->ofdma_tx_mcs_ext[1]);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_nss, "tx_nss",
 			   HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS, "\n");
@@ -3251,7 +3516,10 @@ static inline void htt_print_tx_pdev_rat
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_bw, "ofdma_tx_bw",
 			   HTT_TX_PDEV_STATS_NUM_BW_COUNTERS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_stbc, "tx_stbc",
-			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
+			   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->tx_stbc_ext[0],
+			 htt_stats_buf->tx_stbc_ext[1]);
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_pream, "tx_pream",
 			   HTT_TX_PDEV_STATS_NUM_PREAMBLE_TYPES, "\n");
 
@@ -3260,6 +3528,8 @@ static inline void htt_print_tx_pdev_rat
 			 htt_stats_buf->tx_he_ltf[2],
 			 htt_stats_buf->tx_he_ltf[3]);
 
+	len += scnprintf(buf + len, buf_len - len, "Extra LTF: %u", htt_stats_buf->extra_eht_ltf);
+
 	/* SU GI Stats */
 	for (j = 0; j < HTT_TX_PDEV_STATS_NUM_GI_COUNTERS; j++) {
 		len += scnprintf(buf + len, buf_len - len, "tx_gi[%u] = %s\n",
@@ -3268,10 +3538,10 @@ static inline void htt_print_tx_pdev_rat
 				 htt_stats_buf->tx_gi_ext_2[j][0],
 				 htt_stats_buf->tx_gi_ext_2[j][1]);
 		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_gi[j], NULL,
-				   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->tx_gi_ext[j], NULL,
-				   HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
-
+				   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+		len += scnprintf(buf + len, buf_len - len, "12:%u, 13:%u\n",
+				 htt_stats_buf->tx_gi_ext[j][0],
+				 htt_stats_buf->tx_gi_ext[j][1]);
 	}
 
 	/* AC MU-MIMO GI Stats */
@@ -3280,9 +3550,10 @@ static inline void htt_print_tx_pdev_rat
 				 "ac_mu_mimo_tx_gi[%u] = %s\n",
 				 j, tx_gi[j]);
 		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ac_mu_mimo_tx_gi[j],
-				   NULL, HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_gi_ext[j],
-				   NULL, HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+				   NULL, HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+		len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+				 htt_stats_buf->ax_mu_mimo_tx_gi_ext[j][0],
+				 htt_stats_buf->ax_mu_mimo_tx_gi_ext[j][1]);
 	}
 
 	/* AX MU-MIMO GI Stats */
@@ -3291,9 +3562,10 @@ static inline void htt_print_tx_pdev_rat
 				 "ax_mu_mimo_tx_gi[%u] = %s\n",
 				 j, tx_gi[j]);
 		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_gi[j],
-				   NULL, HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ax_mu_mimo_tx_gi_ext[j],
-				   NULL, HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+				   NULL, HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "  ");
+		len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+				 htt_stats_buf->ax_mu_mimo_tx_gi_ext[j][0],
+				 htt_stats_buf->ax_mu_mimo_tx_gi_ext[j][1]);
 	}
 
 	/* DL OFDMA GI Stats */
@@ -3301,9 +3573,10 @@ static inline void htt_print_tx_pdev_rat
 		len += scnprintf(buf + len, buf_len - len, "ofdma_tx_gi[%u] = %s\n",
 				 j, tx_gi[j]);
 		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_gi[j], NULL,
-				   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-		PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_gi_ext[j],
-				   NULL, HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+				   HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+		len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+				 htt_stats_buf->ofdma_tx_gi_ext[j][0],
+				 htt_stats_buf->ofdma_tx_gi_ext[j][1]);
 	}
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->ofdma_tx_ru_size,
@@ -3421,7 +3694,10 @@ static inline void htt_print_rx_pdev_rat
 			 htt_stats_buf->rssi_in_dbm);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->rx_mcs, "rx_mcs",
-			   HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
+			   HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS, " ");
+	len += scnprintf(buf + len, buf_len - len, "12: %u, 13: %u\n",
+			 htt_stats_buf->rx_mcs_ext[0],
+			 htt_stats_buf->rx_mcs_ext[1]);
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->rx_nss, "rx_nss",
 			   HTT_RX_PDEV_STATS_NUM_SPATIAL_STREAMS, "\n");
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->rx_dcm, "rx_dcm",
@@ -4310,6 +4586,42 @@ htt_print_tx_selfgen_be_stats_tlv(const
 			 htt_stats_buf->be_mu_bar_trigger);
 	len += scnprintf(buf + len, buf_len - len, "be_mu_rts_trigger = %u\n",
 			 htt_stats_buf->be_mu_rts_trigger);
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->combined_be_bsr_trigger_tried,
+			   "combined_be_bsr_trigger_tried",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->combined_be_bsr_trigger_err,
+			   "combined_be_bsr_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->standalone_be_bsr_trigger_tried,
+			   "standalone_be_bsr_trigger_tried",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->standalone_be_bsr_trigger_err,
+			   "standalone_be_bsr_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_be_su_ulofdma_basic_trigger,
+			   "manual_be_su_ulofdma_basic_trigger",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_be_su_ulofdma_basic_trigger_err,
+			   "manual_be_su_ulofdma_basic_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_be_mu_ulofdma_basic_trigger,
+			   "manual_be_mu_ulofdma_basic_trigger",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->manual_be_mu_ulofdma_basic_trigger_err,
+			   "manual_be_mu_ulofdma_basic_trigger_err",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->be_basic_trigger_per_ac,
+			   "be_basic_trigger_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->be_basic_trigger_errors_per_ac,
+			   "be_basic_trigger_errors_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->be_mu_bar_trigger_per_ac,
+			   "be_mu_bar_trigger_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->be_mu_bar_trigger_errors_per_ac,
+			   "be_mu_bar_trigger_errors_per_ac",
+			   HTT_NUM_AC_WMM, "\n");
 
 	stats_req->buf_len = len;
 }
@@ -4491,17 +4803,59 @@ htt_print_be_ul_mimo_user_stats(const vo
 	stats_req->buf_len = len;
 }
 
+static const char *htt_be_rx_ru_size_to_str(enum HTT_RX_PDEV_STATS_BE_RU_SIZE ru_size)
+{
+	switch (ru_size) {
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_26:
+		return "26";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_52:
+		return "52";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_52_26:
+		return "52+26";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_106:
+		return "106";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_106_26:
+		return "106+26";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_242:
+		return "242";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_484:
+		return "484";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_484_242:
+		return "484+242";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996:
+		return "996";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996_484:
+		return "996+484";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996_484_242:
+		return "996+484+242";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996x2:
+		return "996x2";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996x2_484:
+		return "996x2+484";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996x3:
+		return "996x3";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996x3_484:
+		return "996x3+484";
+	case HTT_RX_PDEV_STATS_BE_RU_SIZE_996x4:
+		return "996x4";
+	default:
+		return "unknown";
+	}
+}
+
 static inline void
 htt_print_ul_mumimo_trig_be_stats(const void *tag_buf,
 				  struct debug_htt_stats_req *stats_req)
 {
-	const struct htt_rx_pdev_ul_mumimo_trig_be_stats_tlv *htt_ul_mumimo_trig_be_stats_buf = tag_buf;
+	const struct htt_rx_pdev_ul_mumimo_trig_be_stats_tlv *htt_ulmimo_tri_be_stat_buf = tag_buf;
 	char *rx_gi[HTT_RX_PDEV_STATS_NUM_GI_COUNTERS];
 	u8 i, j;
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
-	u32 mac_id_word = __le32_to_cpu(htt_ul_mumimo_trig_be_stats_buf->mac_id__word);
+	u32 mac_id_word = __le32_to_cpu(htt_ulmimo_tri_be_stat_buf->mac_id__word);
+	u16 index;
+	char str_buf[HTT_MAX_STRING_LEN] = {0};
 
 	for (i = 0; i < HTT_TX_PEER_STATS_NUM_GI_COUNTERS; i++) {
 		rx_gi[i] = kmalloc(HTT_MAX_STRING_LEN, GFP_ATOMIC);
@@ -4515,82 +4869,107 @@ htt_print_ul_mumimo_trig_be_stats(const
 			 u32_get_bits(mac_id_word, HTT_STATS_MAC_ID));
 
 	len += scnprintf(buf + len, buf_len - len, "rx_11be_ul_mumimo = %u\n",
-			 htt_ul_mumimo_trig_be_stats_buf->rx_11be_ul_mumimo);
+			 htt_ulmimo_tri_be_stat_buf->rx_11be_ul_mumimo);
 
 	/* TODO: Check if enough space is present before writing BE MCS Counters */
 	/* MCS -2 and -1 will be printed first */
 
-	len += scnprintf(buf + len, buf_len - len, " -2:%u,-1:%u\n",
-		htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_mcs[HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS-2],
-		htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_mcs[HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS-1]);
-
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_mcs,
-			   "be_ul_mumimo_rx_mcs",
-			   HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 2, "\n\n");
+	index = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
 
-	for (j = 0; j < HTT_RX_PDEV_STATS_NUM_GI_COUNTERS; j++) {
-		len += scnprintf(buf + len, buf_len - len, " -2:%u,-1:%u\n",
-			htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_gi[j][HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS-2],
-			htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_gi[j][HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS-1]);
+	index += snprintf(&str_buf[index], HTT_MAX_STRING_LEN - index, " -2:%u,-1:%u",
+			  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_mcs
+			  [HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 2],
+			  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_mcs
+			  [HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 1]);
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 2; i++)
+		index += snprintf(&str_buf[index],
+				  HTT_MAX_STRING_LEN - index, " %u:%u,", i,
+				  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_mcs[i]);
 
-		PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_gi[j],
-				   NULL, HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS-2, "\n\n");
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_ul_mumimo_rx_mcs = %s\n", str_buf);
+
+	for (j = 0; j < HTT_RX_PDEV_STATS_NUM_GI_COUNTERS; j++) {
+		index = 0;
+		memset(&str_buf[index], 0x0, HTT_MAX_STRING_LEN);
+		index += snprintf(&str_buf[index], HTT_MAX_STRING_LEN - index, " -2:%u,-1:%u",
+				  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_gi[j]
+				  [HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 2],
+				  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_gi[j]
+				  [HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 1]);
+		for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS - 2; i++)
+			index += snprintf(&str_buf[index],
+					  HTT_MAX_STRING_LEN - index, " %u:%u,", i,
+					  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_gi[j][i]);
 
 		len += scnprintf(buf + len, buf_len - len,
-				 "be_ul_mumimo_rx_gi[%u] = %s ", j, rx_gi[j]);
+				 "be_ul_mumimo_rx_gi[%u] = %s\n", j, str_buf);
 	}
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_nss,
-			   "be_ul_mumimo_rx_nss",
-			   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS, "\n\n");
+	index = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+	for (i = 0; i < HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS; i++)
+		index += snprintf(&str_buf[index],
+				  HTT_MAX_STRING_LEN - index, " %u:%u,", i + 1,
+				  htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_nss[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_ul_mumimo_rx_nss = %s\n", str_buf);
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_bw,
+	PRINT_ARRAY_TO_BUF(buf, len, htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_bw,
 			   "be_ul_mumimo_rx_bw",
-			   HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS, "\n\n");
+			   HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS, "\n");
 
 	len += scnprintf(buf + len, buf_len - len,
 			 "be_ul_mumimo_rx_stbc = %u\n",
-			 htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_stbc);
+			 htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_stbc);
 	len += scnprintf(buf + len, buf_len - len,
 			 "be_ul_mumimo_rx_ldpc = %u\n",
-			 htt_ul_mumimo_trig_be_stats_buf->be_ul_mumimo_rx_ldpc);
+			 htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_rx_ldpc);
 
-	for (i = 0; i < HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS; i++) {
-		PRINT_ARRAY_TO_BUF(buf, len,
-				   htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_chain_rssi_in_dbm[i],
-				   NULL, HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS, "\n\n");
+	PRINT_ARRAY_TO_BUF(buf, len, htt_ulmimo_tri_be_stat_buf->rx_ul_mumimo_punctured_mode,
+			   "rx_ul_mumimo_punctured_mode",
+			   HTT_RX_PDEV_STATS_NUM_PUNCTURED_MODE_COUNTERS, "\n");
+
+	for (j = 0; j < HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "be_rx_ul_mumimo_rssi_in_dbm: chain[%u] = %s ",
-				 i, htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_chain_rssi_in_dbm[i]);
+				 "be_rx_ul_mumimo_rssi_in_dbm: chain[%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
+				   htt_ulmimo_tri_be_stat_buf->be_rx_ul_mumimo_chain_rssi_in_dbm[j],
+				   HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
-	for (i = 0; i < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; i++) {
-		PRINT_ARRAY_TO_BUF(buf, len,
-				   htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_target_rssi[i],
-				   NULL, HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS, "\n\n");
+	for (j = 0; j < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "be_rx_ul_mumimo_target_rssi: user[%u] = %s ",
-				 i, htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_target_rssi[i]);
+				 "be_rx_ul_mumimo_target_rssi: user[%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
+				   htt_ulmimo_tri_be_stat_buf->be_rx_ul_mumimo_target_rssi[j],
+				   HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
-	for (i = 0; i < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; i++) {
-		PRINT_ARRAY_TO_BUF(buf, len,
-				htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_fd_rssi[i],
-				NULL, HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS, "\n\n");
+	for (j = 0; j < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "be_rx_ul_mumimo_fd_rssi: user[%u] = %s ",
-				 i, htt_ul_mumimo_trig_be_stats_buf->be_rx_ul_mumimo_fd_rssi[i]);
+				 "be_rx_ul_mumimo_fd_rssi: user[%u] =  ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
+				   htt_ulmimo_tri_be_stat_buf->be_rx_ul_mumimo_fd_rssi[j],
+				   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
-	for (i = 0; i < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; i++) {
-		PRINT_ARRAY_TO_BUF(buf, len,
-				htt_ul_mumimo_trig_be_stats_buf->be_rx_ulmumimo_pilot_evm_dB_mean[i],
-				NULL, HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS, "\n\n");
+	for (j = 0; j < HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "be_rx_ulmumimo_pilot_evm_dB_mean: user [%u] = %s ",
-				 i, htt_ul_mumimo_trig_be_stats_buf->be_rx_ulmumimo_pilot_evm_dB_mean[i]);
+				 "be_rx_ulmumimo_pilot_evm_dB_mean: user [%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
+				   htt_ulmimo_tri_be_stat_buf->be_rx_ulmumimo_pilot_evm_dB_mean[j],
+				   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
-	len += scnprintf(buf + len, buf_len - len, "\n");
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_ul_mumimo_basic_trigger_rx_qos_null_only = %u\n",
+			 htt_ulmimo_tri_be_stat_buf->be_ul_mumimo_basic_trigger_rx_qos_null_only);
 
 	stats_req->buf_len = len;
 fail:
@@ -4609,6 +4988,7 @@ htt_print_be_ul_ofdma_trigger_stats(cons
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
 	u32 mac_id_word = __le32_to_cpu(htt_trigger_stats_buf->mac_id__word);
+	char str_buf[HTT_MAX_STRING_LEN] = {0};
 
 	for (i = 0; i < HTT_TX_PEER_STATS_NUM_GI_COUNTERS; i++) {
 		rx_gi[i] = kmalloc(HTT_MAX_STRING_LEN, GFP_ATOMIC);
@@ -4630,33 +5010,54 @@ htt_print_be_ul_ofdma_trigger_stats(cons
 	for (j = 0; j < HTT_RX_PDEV_STATS_NUM_GI_COUNTERS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
 				 "be_ul_ofdma_rx_gi[%u] = %s ", j, rx_gi[j]);
-		for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS; i++) {
+		for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_MCS_COUNTERS; i++)
 			len += scnprintf(buf + len, buf_len - len, " %u:%u", i,
 					 htt_trigger_stats_buf->be_ul_ofdma_rx_gi[j][i]);
-		}
+
 		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_ul_ofdma_rx_nss,
-			   "be_ul_ofdma_rx_nss",
-			   HTT_RX_PDEV_STATS_NUM_SPATIAL_STREAMS, "\n");
+	j = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++)
+		j += snprintf(&str_buf[j],
+			      HTT_MAX_STRING_LEN - j,
+			      " %u:%u,", i + 1,
+			      htt_trigger_stats_buf->be_ul_ofdma_rx_nss[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_ul_ofdma_rx_nss = %s\n", str_buf);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_ul_ofdma_rx_bw,
 			   "be_ul_ofdma_rx_bw",
 			   HTT_RX_PDEV_STATS_NUM_BE_BW_COUNTERS, "\n");
 
 	len += scnprintf(buf + len, buf_len - len, "be_ul_ofdma_rx_stbc = %u\n",
-			htt_trigger_stats_buf->be_ul_ofdma_rx_stbc);
+			 htt_trigger_stats_buf->be_ul_ofdma_rx_stbc);
 	len += scnprintf(buf + len, buf_len - len, "be_ul_ofdma_rx_ldpc = %u\n",
-			htt_trigger_stats_buf->be_ul_ofdma_rx_ldpc);
+			 htt_trigger_stats_buf->be_ul_ofdma_rx_ldpc);
+
+	j = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS; i++)
+		j += snprintf(&str_buf[j], HTT_MAX_STRING_LEN - j,
+			      " %s:%u,", htt_be_rx_ru_size_to_str(i),
+			      htt_trigger_stats_buf->be_rx_ulofdma_data_ru_size_ppdu[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_rx_ulofdma_data_ru_size_ppdu = %s\n", str_buf);
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_rx_ulofdma_data_ru_size_ppdu,
-			   "be_rx_ulofdma_data_ru_size_ppdu",
-			   HTT_RX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS, "\n");
-
-	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_rx_ulofdma_non_data_ru_size_ppdu,
-			   "be_rx_ulofdma_non_data_ru_size_ppdu",
-			   HTT_RX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS, "\n");
+	j = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS; i++)
+		j += snprintf(&str_buf[j], HTT_MAX_STRING_LEN - j,
+			      " %s:%u,", htt_be_rx_ru_size_to_str(i),
+			      htt_trigger_stats_buf->be_rx_ulofdma_non_data_ru_size_ppdu[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_rx_ulofdma_non_data_ru_size_ppdu = %s\n", str_buf);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_uplink_sta_aid,
 			   "be_rx_rssi_track_sta_aid",
@@ -4673,7 +5074,19 @@ htt_print_be_ul_ofdma_trigger_stats(cons
 	PRINT_ARRAY_TO_BUF(buf, len, htt_trigger_stats_buf->be_uplink_sta_power_headroom,
 			   "be_rx_sta_power_headroom",
 			   HTT_RX_UL_MAX_UPLINK_RSSI_TRACK, "\n");
-
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_ul_ofdma_basic_trigger_rx_qos_null_only = %u\n",
+			 htt_trigger_stats_buf->be_ul_ofdma_basic_trigger_rx_qos_null_only);
+	len += scnprintf(buf + len, buf_len - len, "ul_mlo_send_qdepth_params_count = %u\n",
+			 htt_trigger_stats_buf->ul_mlo_send_qdepth_params_count);
+	len += scnprintf(buf + len, buf_len - len, "ul_mlo_proc_qdepth_params_count = %u\n",
+			 htt_trigger_stats_buf->ul_mlo_proc_qdepth_params_count);
+	len += scnprintf(buf + len, buf_len - len,
+			 "ul_mlo_proc_accepted_qdepth_params_count = %u\n",
+			 htt_trigger_stats_buf->ul_mlo_proc_accepted_qdepth_params_count);
+	len += scnprintf(buf + len, buf_len - len,
+			 "ul_mlo_proc_discarded_qdepth_params_count = %u\n",
+			 htt_trigger_stats_buf->ul_mlo_proc_discarded_qdepth_params_count);
 	len += scnprintf(buf + len, buf_len - len, "\n");
 
 	stats_req->buf_len = len;
@@ -4857,6 +5270,17 @@ htt_print_tx_selfgen_be_err_stats_tlv(co
 	len += scnprintf(buf + len, buf_len - len,
 			 "be_mu_rts_trigger_err = %u\n",
 			 htt_stats_buf->be_mu_rts_trigger_err);
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_basic_trigger_partial_resp = %u\n",
+			 htt_stats_buf->be_basic_trigger_partial_resp);
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_bsr_trigger_partial_resp = %u\n",
+			 htt_stats_buf->be_bsr_trigger_partial_resp);
+	len += scnprintf(buf + len, buf_len - len,
+			 "be_mu_bar_trigger_partial_resp = %u\n",
+			 htt_stats_buf->be_mu_bar_trigger_partial_resp);
+
+	stats_req->buf_len = len;
 }
 
 static inline void htt_print_latency_prof_stats_tlv_v(const u32 *tag_buf, u8 *data)
@@ -4868,21 +5292,31 @@ static inline void htt_print_latency_pro
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
+	u32 page_fault_avg = 0;
 
 	if (htt_stats_buf->print_header == 1) {
 		len += scnprintf(buf + len, buf_len - len,
-				   "HTT_STATS_LATENCY_PROF_TLV:\n");
+				 "HTT_STATS_LATENCY_PROF_TLV:\n");
 		len += scnprintf(buf + len, buf_len - len,
-				   "|  prof_name\t| cnt\t| min\t| max\t| last\t| tot\t| avg \t| hist_intvl\t| hist\t|\n");
-	}
-
-	len += scnprintf(buf + len, buf_len - len,
-			   "|%s| %8u| %8u| %8u| %8u| %8u| %8u| %8u| %4u:%4u:%4u|\n",
-			   htt_stats_buf->latency_prof_name, htt_stats_buf->cnt,
-			   htt_stats_buf->min, htt_stats_buf->max, htt_stats_buf->last,
-			   htt_stats_buf->tot, htt_stats_buf->avg,
-			   htt_stats_buf->hist_intvl, htt_stats_buf->hist[0],
-			   htt_stats_buf->hist[1], htt_stats_buf->hist[2]);
+				 "|%-32s|%8s|%15s|%8s|%8s|%8s|%8s|%15s|%26s|%8s|%8s|%8s|%10s|%17s|%6s|\n",
+				 "prof_name", "cnt", "min", "max", "last", "tot", "avg",
+				 "hist_intvl", "hist", "pf_max", "pf_avg", "pf_tot", "ignoredCnt",
+				 "intHist", "intMax");
+	}
+	if (htt_stats_buf->cnt)
+		page_fault_avg = htt_stats_buf->page_fault_total / htt_stats_buf->cnt;
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "|%-32s|%8u|%15u|%8u|%8u|%8u|%8u|%15u|%8u:%8u:%8u|%8u|%8u|%8u|%10u|%5u:%5u:%5u|%6u|\n",
+			 htt_stats_buf->latency_prof_name, htt_stats_buf->cnt,
+			 htt_stats_buf->min, htt_stats_buf->max, htt_stats_buf->last,
+			 htt_stats_buf->tot, htt_stats_buf->avg,
+			 htt_stats_buf->hist_intvl, htt_stats_buf->hist[0],
+			 htt_stats_buf->hist[1], htt_stats_buf->hist[2],
+			 htt_stats_buf->page_fault_max, page_fault_avg,
+			 htt_stats_buf->page_fault_total, htt_stats_buf->ignored_latency_count,
+			 htt_stats_buf->interrupts_hist[0], htt_stats_buf->interrupts_hist[1],
+			 htt_stats_buf->interrupts_hist[2], htt_stats_buf->interrupts_max);
 
 	stats_req->buf_len = len;
 }
@@ -4935,7 +5369,9 @@ static inline void htt_print_ul_mumimo_t
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
 	u32 mac_id = __le32_to_cpu(htt_ul_mumimo_trig_stats_buf->mac_id__word);
-	u8 j;
+	u8 i, j;
+	u16 index;
+	char str_buf[HTT_MAX_STRING_LEN] = {0};
 
 	len += scnprintf(buf + len, buf_len - len,
 			 "HTT_RX_PDEV_UL_MUMIMO_TRIG_STATS_TLV:\n");
@@ -4945,72 +5381,111 @@ static inline void htt_print_ul_mumimo_t
 	len += scnprintf(buf + len, buf_len - len,
 			 "rx_11ax_ul_mumimo = %u\n",
 			 htt_ul_mumimo_trig_stats_buf->rx_11ax_ul_mumimo);
+	index = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS; i++)
+		index += snprintf(&str_buf[index],
+				  HTT_MAX_STRING_LEN - index,
+				  " %u:%u,", i, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_mcs[i]);
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_mcs,
-			   "ul_mumimo_rx_mcs", HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_mcs_ext,
-			   "ul_mumimo_rx_mcs_ext",
-			   HTT_RX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS; i++)
+		index += snprintf(&str_buf[index],
+				  HTT_MAX_STRING_LEN - index,
+				  " %u:%u,", i + HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS,
+				  htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_mcs_ext[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "ul_mumimo_rx_mcs = %s\n", str_buf);
 
 	for (j = 0; j < HTT_RX_PDEV_STATS_NUM_GI_COUNTERS; j++) {
-		len += scnprintf(buf + len, buf_len - len, "\nul_mumimo_rx_gi[%u] = ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
-				   htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_gi[j],
-				   NULL, HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS, "\n");
-		len += scnprintf(buf + len, buf_len - len, ", ");
-		PRINT_ARRAY_TO_BUF(buf, len,
-				   htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_gi_ext[j],
-				   NULL, HTT_RX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS, "\n");
+		index = 0;
+		memset(&str_buf[index], 0x0, HTT_MAX_STRING_LEN);
+		for (i = 0; i < HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS; i++)
+			index += snprintf(&str_buf[index],
+					  HTT_MAX_STRING_LEN - index, " %u:%u,",
+					  i, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_gi[j][i]);
+
+		for (i = 0; i < HTT_RX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS; i++)
+			index += snprintf(&str_buf[index],
+					  HTT_MAX_STRING_LEN - index, " %u:%u,",
+					  i + HTT_RX_PDEV_STATS_NUM_MCS_COUNTERS,
+					  htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_gi_ext[j][i]);
+
+		len += scnprintf(buf + len, buf_len - len,
+				 "ul_mumimo_rx_gi[%u] = %s\n", j, str_buf);
 	}
 
-	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_nss,
-			   "ul_mumimo_rx_nss",
-			   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS, "\n");
+	index = 0;
+	memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+	for (i = 0; i < HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS; i++)
+		index += snprintf(&str_buf[index],
+				  HTT_MAX_STRING_LEN - index, " %u:%u,", i + 1,
+				  htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_nss[i]);
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "ul_mumimo_rx_nss = %s\n", str_buf);
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_bw,
 			   "ul_mumimo_rx_bw",
 			   HTT_RX_PDEV_STATS_NUM_BW_COUNTERS, "\n");
 
-	len += scnprintf(buf + len, buf_len - len, "\nul_mumimo_rx_stbc = %u",
+	for (i = 0; i < HTT_RX_PDEV_STATS_NUM_REDUCED_CHAN_TYPES; i++) {
+		index = 0;
+		memset(str_buf, 0x0, HTT_MAX_STRING_LEN);
+		for (j = 0; j < HTT_RX_PDEV_STATS_NUM_BW_COUNTERS; j++)
+			index += snprintf(&str_buf[index],
+					  HTT_MAX_STRING_LEN - index, " %u:%u,", j,
+					  htt_ul_mumimo_trig_stats_buf->reduced_ul_mumimo_rx_bw
+					  [i][j]);
+
+		len += scnprintf(buf + len, buf_len - len, "%s = %s\n",
+				 i == 0 ? "half_ul_mumimo_rx_bw" : "quarter_ul_mumimo_rx_bw",
+				 str_buf);
+	}
+
+	len += scnprintf(buf + len, buf_len - len, "ul_mumimo_rx_stbc = %u\n",
 			 htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_stbc);
-	len += scnprintf(buf + len, buf_len - len, "\nul_mumimo_rx_ldpc = %u",
+	len += scnprintf(buf + len, buf_len - len, "ul_mumimo_rx_ldpc = %u\n",
 			 htt_ul_mumimo_trig_stats_buf->ul_mumimo_rx_ldpc);
 
 	for (j = 0; j < HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "\nrx_ul_mumimo_rssi_in_dbm: chain[%u] = ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
+				 "rx_ul_mumimo_rssi_in_dbm: chain[%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
 				   htt_ul_mumimo_trig_stats_buf->rx_ul_mumimo_chain_rssi_in_dbm[j],
-				   NULL, HTT_RX_PDEV_STATS_TOTAL_BW_COUNTERS, "\n");
+				   HTT_RX_PDEV_STATS_TOTAL_BW_COUNTERS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
 	for (j = 0; j < HTT_TX_PDEV_STATS_NUM_UL_MUMIMO_USER_STATS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				   "\nrx_ul_mumimo_target_rssi: user[%u] = ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
+				 "rx_ul_mumimo_target_rssi: user[%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
 				   htt_ul_mumimo_trig_stats_buf->rx_ul_mumimo_target_rssi[j],
-				   NULL, HTT_RX_PDEV_STATS_NUM_BW_COUNTERS, "\n");
+				   HTT_RX_PDEV_STATS_NUM_BW_COUNTERS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
 	for (j = 0; j < HTT_TX_PDEV_STATS_NUM_UL_MUMIMO_USER_STATS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "\nrx_ul_mumimo_fd_rssi: user[%u] =  ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
+				 "rx_ul_mumimo_fd_rssi: user[%u] =  ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
 				   htt_ul_mumimo_trig_stats_buf->rx_ul_mumimo_fd_rssi[j],
-				   NULL, HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS,
-				   "\n");
+				   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
 	for (j = 0; j < HTT_TX_PDEV_STATS_NUM_UL_MUMIMO_USER_STATS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
-				 "\nrx_ulmumimo_pilot_evm_db_mean: user [%u] = ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
+				 "rx_ulmumimo_pilot_evm_db_mean: user [%u] = ", j);
+		CHAIN_ARRAY_TO_BUF(buf, len,
 				   htt_ul_mumimo_trig_stats_buf->rx_ulmumimo_pilot_evm_db_mean[j],
-				   NULL, HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS,
-				   "\n");
+				   HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
-
-	len += scnprintf(buf + len, buf_len - len, "\n");
+	len += scnprintf(buf + len, buf_len - len,
+			 "ul_mumimo_basic_trigger_rx_qos_null_only = %u\n",
+			 htt_ul_mumimo_trig_stats_buf->ul_mumimo_basic_trigger_rx_qos_null_only);
 
 	stats_req->buf_len = len;
 }
@@ -5359,6 +5834,78 @@ htt_print_pdev_tx_rate_txbf_stats_tlv(co
 	stats_req->buf_len = len;
 }
 
+static const char *htt_ax_tx_ru_size_to_str(enum HTT_TX_PDEV_STATS_AX_RU_SIZE ru_size)
+{
+	switch (ru_size) {
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_26:
+		return "26";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_52:
+		return "52";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_106:
+		return "106";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_242:
+		return "242";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_484:
+		return "484";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_996:
+		return "996";
+	case HTT_TX_PDEV_STATS_AX_RU_SIZE_996x2:
+		return "996x2";
+	default:
+		return "unknown";
+	}
+}
+
+static const char *htt_be_tx_ru_size_to_str(enum HTT_RX_PDEV_STATS_BE_RU_SIZE ru_size)
+{
+	switch (ru_size) {
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_26:
+		return "26";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_52:
+		return "52";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_52_26:
+		return "52+26";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_106:
+		return "106";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_106_26:
+		return "106+26";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_242:
+		return "242";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_484:
+		return "484";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_484_242:
+		return "484+242";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996:
+		return "996";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996_484:
+		return "996+484";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996_484_242:
+		return "996+484+242";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996x2:
+		return "996x2";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996x2_484:
+		return "996x2+484";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996x3:
+		return "996x3";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996x3_484:
+		return "996x3+484";
+	case HTT_TX_PDEV_STATS_BE_RU_SIZE_996x4:
+		return "996x4";
+	default:
+		return "unknown";
+	}
+}
+
+static const char *htt_stats_tx_ru_size_to_str(enum htt_stats_ru_type ru_type, u32 ru_size)
+{
+	if (ru_type == HTT_STATS_RU_TYPE_SINGLE_RU_ONLY)
+		return htt_ax_tx_ru_size_to_str(ru_size);
+	else if (ru_type == HTT_STATS_RU_TYPE_SINGLE_AND_MULTI_RU)
+		return htt_be_tx_ru_size_to_str(ru_size);
+	else
+		return "unknown";
+}
+
 static inline void
 htt_print_tx_per_rate_stats_tlv(const void *tag_buf,
 				struct debug_htt_stats_req *stats_req)
@@ -5367,27 +5914,49 @@ htt_print_tx_per_rate_stats_tlv(const vo
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
-	char mode_prefix[][3] = {"su", "mu"};
+	char mode_prefix[][8] = {"su", "mu", "ofdma", "ulmu", "ulofdma"};
 	int i = 0, j = 0;
+	u32 ru_size_cnt = 0;
 
 	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_DLSU) {
 		len += scnprintf(buf + len, buf_len - len, "HTT_TX_PER_STATS:\n");
-		len += scnprintf(buf + len, buf_len - len, "PER_STATS_SU:\n");
+		len += scnprintf(buf + len, buf_len - len, "\nPER_STATS_SU:\n");
 	} else if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_DLMUMIMO) {
 		len += scnprintf(buf + len, buf_len - len, "\nPER_STATS_DL_MUMIMO:\n");
 		j = 1;
+	} else if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_DLOFDMA) {
+		len += scnprintf(buf + len, buf_len - len, "\n\nPER_STATS_DL_OFDMA:\n");
+		j = 2;
+		if (htt_stats_buf->ru_type == HTT_STATS_RU_TYPE_SINGLE_RU_ONLY)
+			ru_size_cnt = HTT_TX_PDEV_STATS_NUM_AX_RU_SIZE_COUNTERS;
+		else if (htt_stats_buf->ru_type == HTT_STATS_RU_TYPE_SINGLE_AND_MULTI_RU)
+			ru_size_cnt = HTT_TX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS;
+	} else if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO) {
+		len += scnprintf(buf + len, buf_len - len, "\nHTT_RX_PER_STATS:\n");
+		len += scnprintf(buf + len, buf_len - len, "\nPER_STATS_UL_MUMIMO:\n");
+		j = 3;
+	} else if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA) {
+		len += scnprintf(buf + len, buf_len - len, "\nPER_STATS_UL_OFDMA:\n");
+		j = 4;
+		if (htt_stats_buf->ru_type == HTT_STATS_RU_TYPE_SINGLE_RU_ONLY)
+			ru_size_cnt = HTT_TX_PDEV_STATS_NUM_AX_RU_SIZE_COUNTERS;
+		else if (htt_stats_buf->ru_type == HTT_STATS_RU_TYPE_SINGLE_AND_MULTI_RU)
+			ru_size_cnt = HTT_TX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS;
 	} else {
 		return;
 	}
 
 	len += scnprintf(buf + len, buf_len - len, "\nPER per BW:\n");
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len, "data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s = ", mode_prefix[j]);
 
-	len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++) {
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_bw[i].ppdus_tried);
-	}
 	/*
 	 * Now i = HTT_TX_PDEV_STATS_NUM_BW_COUNTERS.
 	 * Adding 320 MHz ppdus tried.
@@ -5395,12 +5964,17 @@ htt_print_tx_per_rate_stats_tlv(const vo
 	len += scnprintf(buf + len, buf_len - len,
 			 " %u:%u\n", i, htt_stats_buf->per_bw320.ppdus_tried);
 
-	len += scnprintf(buf + len, buf_len - len, "ppdus_ack_failed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++) {
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len, "non_data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len,
+				 "ppdus_ack_failed_%s = ", mode_prefix[j]);
+
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_bw[i].ppdus_ack_failed);
-	}
 	/*
 	 * Now i = HTT_TX_PDEV_STATS_NUM_BW_COUNTERS.
 	 * Adding 320 MHz ppdus ack failed.
@@ -5408,12 +5982,11 @@ htt_print_tx_per_rate_stats_tlv(const vo
 	len += scnprintf(buf + len, buf_len - len,
 			 " %u:%u\n", i, htt_stats_buf->per_bw320.ppdus_ack_failed);
 
-	len += scnprintf(buf + len, buf_len - len, "mpdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "mpdus_tried_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_bw[i].mpdus_tried);
-	}
 	/*
 	 * Now i = HTT_TX_PDEV_STATS_NUM_BW_COUNTERS.
 	 * Adding 320 MHz mpdus tried.
@@ -5422,12 +5995,11 @@ htt_print_tx_per_rate_stats_tlv(const vo
 			 " %u:%u\n", i, htt_stats_buf->per_bw320.mpdus_tried);
 
 
-	len += scnprintf(buf + len, buf_len - len, "mpdus_failed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "mpdus_failed_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_BW_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_bw[i].mpdus_failed);
-	}
 	/*
 	 * Now i = HTT_TX_PDEV_STATS_NUM_BW_COUNTERS.
 	 * Adding 320 MHz mpdus failed.
@@ -5437,65 +6009,127 @@ htt_print_tx_per_rate_stats_tlv(const vo
 
 	len += scnprintf(buf + len, buf_len - len, "\nPER per NSS:\n");
 
-	len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++) {
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len, "data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s = ", mode_prefix[j]);
+
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_nss[i].ppdus_tried);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nppdus_ack_fialed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++) {
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nnon_data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nppdus_ack_failed_%s = ", mode_prefix[j]);
+
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_nss[i].ppdus_ack_failed);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nmpdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "\nmpdus_tried_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_nss[i].mpdus_tried);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nmpdus_failed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "\nmpdus_failed_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_PDEV_STATS_NUM_SPATIAL_STREAMS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_nss[i].mpdus_failed);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nPER per Mcs:\n");
+	len += scnprintf(buf + len, buf_len - len, "\n\nPER per Mcs:\n");
+
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len, "data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s = ", mode_prefix[j]);
 
-	len += scnprintf(buf + len, buf_len - len, "ppdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++) {
+	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_mcs[i].ppdus_tried);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nppdus_ack_failed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++) {
+	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULMUMIMO ||
+	    htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nnon_data_ppdus_%s = ", mode_prefix[j]);
+	else
+		len += scnprintf(buf + len, buf_len - len,
+				 "\nppdus_ack_failed_%s = ", mode_prefix[j]);
+
+	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_mcs[i].ppdus_ack_failed);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nmpdus_tried_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "\nmpdus_tried_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_mcs[i].mpdus_tried);
-	}
 
-	len += scnprintf(buf + len, buf_len - len, "\nmpdus_failed_%s", mode_prefix[j]);
-	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++) {
+	len += scnprintf(buf + len, buf_len - len, "\nmpdus_failed_%s = ", mode_prefix[j]);
+	for (i = 0; i < HTT_TX_TXBF_RATE_STATS_NUM_MCS_COUNTERS; i++)
 		len += scnprintf(buf + len, buf_len - len,
 				 " %u:%u ",
 				 i, htt_stats_buf->per_mcs[i].mpdus_failed);
+
+	if ((htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_DLOFDMA ||
+	     htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA) &&
+	    htt_stats_buf->ru_type != HTT_STATS_RU_TYPE_INVALID) {
+		len += scnprintf(buf + len, buf_len - len, "\n\nPER per RU:\n");
+
+		if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+			len += scnprintf(buf + len, buf_len - len,
+					 "\ndata_ppdus_%s = ", mode_prefix[j]);
+		else
+			len += scnprintf(buf + len, buf_len - len,
+					 "\nppdus_tried_%s = ", mode_prefix[j]);
+
+		for (i = 0; i < ru_size_cnt; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %s:%u ",
+					 htt_stats_tx_ru_size_to_str(htt_stats_buf->ru_type, i),
+					 htt_stats_buf->per_ru[i].ppdus_tried);
+
+		if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_ULOFDMA)
+			len += scnprintf(buf + len, buf_len - len,
+					 "\nnon_data_ppdus_%s = ", mode_prefix[j]);
+		else
+			len += scnprintf(buf + len, buf_len - len,
+					 "\nppdus_ack_failed_%s = ", mode_prefix[j]);
+
+		for (i = 0; i < ru_size_cnt; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %s:%u ",
+					 htt_stats_tx_ru_size_to_str(htt_stats_buf->ru_type, i),
+					 htt_stats_buf->per_ru[i].ppdus_ack_failed);
+
+		len += scnprintf(buf + len, buf_len - len, "\nmpdus_tried_%s = ", mode_prefix[j]);
+		for (i = 0; i < ru_size_cnt; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %s:%u ",
+					 htt_stats_tx_ru_size_to_str(htt_stats_buf->ru_type, i),
+					 htt_stats_buf->per_ru[i].mpdus_tried);
+
+		len += scnprintf(buf + len, buf_len - len, "\nmpdus_failed_%s = ", mode_prefix[j]);
+		for (i = 0; i < ru_size_cnt; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %s:%u ",
+					 htt_stats_tx_ru_size_to_str(htt_stats_buf->ru_type, i),
+					 htt_stats_buf->per_ru[i].mpdus_failed);
 	}
 
-	len += scnprintf(buf + len, buf_len - len, "\n");
+	len += scnprintf(buf + len, buf_len - len, "\n\n");
 
 	if (htt_stats_buf->rc_mode == HTT_STATS_RC_MODE_DLMUMIMO) {
 		len += scnprintf(buf + len, buf_len - len, "last_probed_bw  = %u\n",
@@ -5504,6 +6138,12 @@ htt_print_tx_per_rate_stats_tlv(const vo
 				 htt_stats_buf->last_probed_nss);
 		len += scnprintf(buf + len, buf_len - len, "last_probed_mcs = %u\n",
 				 htt_stats_buf->last_probed_mcs);
+
+		len += scnprintf(buf + len, buf_len - len, "MU Probe count per RC MODE = ");
+		for (i = 0; i < HTT_RC_MODE_2D_COUNT; i++)
+			len += scnprintf(buf + len, buf_len - len,
+					 " %u:%u ",
+					 i, htt_stats_buf->probe_cnt_per_rcmode[i]);
 	}
 
 	stats_req->buf_len = len;
@@ -6289,7 +6929,7 @@ htt_print_phy_tpc_stats_tlv(const void *
 	u8 *buf = stats_req->buf;
 	u32 len = stats_req->buf_len;
 	u32 buf_len = ATH12K_HTT_STATS_BUF_SIZE;
-	int i;
+	int i, j, k;
 
 	len += scnprintf(buf + len, buf_len - len, "HTT_PHY_TPC_STATS_TLV:\n");
 	len += scnprintf(buf + len, buf_len - len,
@@ -6359,6 +6999,35 @@ htt_print_phy_tpc_stats_tlv(const void *
 	}
 
 	len += scnprintf(buf + len, buf_len - len,
+			 "\ntpc_stats : ctl_array_gain_cap_ext2_enabled = %u",
+			 HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_GET
+			 (htt_stats_phy_tpc_stats_buf->ctl_args));
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "\ntpc_stats : ctl_flag = %u",
+			 HTT_PHY_TPC_STATS_CTL_FLAG_GET
+			 (htt_stats_phy_tpc_stats_buf->ctl_args));
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "\ntpc_stats : ctl_array_gain_cap_ext2_ctl_region_grp = %u",
+			 HTT_PHY_TPC_STATS_CTL_REGION_GRP_GET
+			 (htt_stats_phy_tpc_stats_buf->ctl_args));
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "\ntpc_stats : ctl_sub_band_index = %u",
+			 HTT_PHY_TPC_STATS_SUB_BAND_INDEX_GET
+			 (htt_stats_phy_tpc_stats_buf->ctl_args));
+
+	k = 0;
+	for (i = 0; i < HTT_STATS_MAX_CHAINS; i++) {
+		for (j = 0; j <= i; j++, k++) {
+			len += scnprintf(buf + len, buf_len - len,
+					 "\ntpc_stats : ctl_array_gain_cap[ntx:%d][nss:%d] =%d",
+					 i + 1, j + 1,
+					 htt_stats_phy_tpc_stats_buf->array_gain_cap[k]);
+		}
+	}
+	len += scnprintf(buf + len, buf_len - len,
 			 "\n=========================================== \n");
 	stats_req->buf_len = len;
 }
@@ -6799,6 +7468,51 @@ htt_print_pdev_sched_algo_ofdma_stats_tl
 			   "chan_acc_lat_based_dlofdma_monitoring_count",
 			   HTT_NUM_AC_WMM, "\n");
 
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->downgrade_to_dl_su_ru_alloc_fail,
+			   "downgrade_to_dl_su_ru_alloc_fail",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->candidate_list_single_user_disable_ofdma,
+			   "candidate_list_single_user_disable_ofdma",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->dl_cand_list_dropped_high_ul_qos_weight,
+			   "dl_cand_list_dropped_high_ul_qos_weight",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->ax_dlofdma_disabled_due_to_pipelining,
+			   "ax_dlofdma_disabled_due_to_pipelining",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->dlofdma_disabled_su_only_eligible,
+			   "dlofdma_disabled_su_only_eligible",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->dlofdma_disabled_consec_no_mpdus_tried,
+			   "dlofdma_disabled_consec_no_mpdus_tried",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->dlofdma_disabled_consec_no_mpdus_success,
+			   "dlofdma_disabled_consec_no_mpdus_success",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->txbf_ofdma_ineligibility_stat,
+			   "sched_ofdma_txbf_ineligibility",
+			   HTT_STATS_SCHED_OFDMA_TXBF_INELIGIBILITY_MAX, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->avg_chan_acc_lat_hist,
+			   "avg_chan_acc_lat_hist",
+			   HTT_MAX_NUM_CHAN_ACC_LAT_INTR, "\n");
+
 	len += scnprintf(buf + len, buf_len - len,
 			 "=========================================== \n");
 	stats_req->buf_len = len;
@@ -7014,6 +7728,16 @@ htt_print_tx_pdev_mu_edca_params_stats_t
 			   HTT_NUM_AC_WMM, "\n");
 
 	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->mumimo_relaxed_mu_edca,
+			   "mumimo_relaxed_mu_edca",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
+			   htt_stats_buf->muofdma_aggressive_mu_edca,
+			   "muofdma_aggressive_mu_edca",
+			   HTT_NUM_AC_WMM, "\n");
+
+	PRINT_ARRAY_TO_BUF(buf, len,
 			   htt_stats_buf->muofdma_relaxed_mu_edca,
 			   "muofdma_relaxed_mu_edca",
 			   HTT_NUM_AC_WMM, "\n");
@@ -7435,6 +8159,31 @@ static inline void htt_print_phy_counter
 				 "rx_ota_err_cnt[%d] = %u\n",
 				 i, htt_stats_buf->rx_ota_err_cnt[i]);
 
+	for (i = 0; i < HTT_MAX_RX_PKT_CNT_EXT; i++)
+		len += scnprintf(buf + len, buf_len - len,
+				 "rx_pkt_cnt_ext[%d] =  %u\n",
+				 i, htt_stats_buf->rx_pkt_cnt_ext[i]);
+
+	for (i = 0; i < HTT_MAX_RX_PKT_CRC_PASS_CNT_EXT; i++)
+		len += scnprintf(buf + len, buf_len - len,
+				 "rx_pkt_crc_pass_cnt_ext[%d] =  %u\n",
+				 i, htt_stats_buf->rx_pkt_crc_pass_cnt_ext[i]);
+
+	for (i = 0; i < HTT_MAX_RX_PKT_MU_CNT; i++)
+		len += scnprintf(buf + len, buf_len - len,
+				 "rx_pkt_mu_cnt[%d] =  %u\n",
+				 i, htt_stats_buf->rx_pkt_mu_cnt[i]);
+
+	for (i = 0; i < HTT_MAX_TX_PKT_CNT; i++)
+		len += scnprintf(buf + len, buf_len - len,
+				"tx_pkt_cnt[%d] =  %u\n",
+				i, htt_stats_buf->tx_pkt_cnt[i]);
+
+	for (i = 0; i < HTT_MAX_PHY_TX_ABORT_CNT; i++)
+		len += scnprintf(buf + len, buf_len - len,
+				 "phy_tx_abort_cnt[%d] =  %u\n",
+				 i, htt_stats_buf->phy_tx_abort_cnt[i]);
+
 	len += scnprintf(buf + len, buf_len - len,
 			 "=================================================\n");
 
@@ -7520,6 +8269,9 @@ static inline void htt_print_phy_reset_c
 			 "reset_counts : rf_mode_switch_phy_off_cnt = %u\n",
 			 htt_stats_buf->rf_mode_switch_phy_off_cnt);
 	len += scnprintf(buf + len, buf_len - len,
+			 "reset_counts : temperature_recal_cnt = %u\n",
+			 htt_stats_buf->temperature_recal_cnt);
+	len += scnprintf(buf + len, buf_len - len,
 			 "=================================================\n");
 
 	stats_req->buf_len = len;
@@ -7614,6 +8366,36 @@ static inline void htt_print_phy_reset_s
 			 "reset_stats : rxdesense_thresh_sw = 0x%x\n",
 			 htt_stats_buf->rxdesense_thresh_sw);
 	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : phy_bw_code = %u\n",
+			 htt_stats_buf->phy_bw_code);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : phy_rate_mode = %u\n",
+			 htt_stats_buf->phy_rate_mode);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : phy_band_code = %u\n",
+			 htt_stats_buf->phy_band_code);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : phy_vreg_base = 0x%x\n",
+			 htt_stats_buf->phy_vreg_base);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : phy_vreg_base_ext = 0x%x\n",
+			 htt_stats_buf->phy_vreg_base_ext);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : cur_table_index = %u\n",
+			 htt_stats_buf->cur_table_index);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : whal_config_flag = 0x%x\n",
+			 htt_stats_buf->whal_config_flag);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : nfcal_iteration_count_home = %u\n",
+			 htt_stats_buf->nfcal_iteration_counts[0]);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : nfcal_iteration_count_scan = %u\n",
+			 htt_stats_buf->nfcal_iteration_counts[1]);
+	len += scnprintf(buf + len, buf_len - len,
+			 "reset_stats : nfcal_iteration_count_periodic = %u\n",
+			 htt_stats_buf->nfcal_iteration_counts[2]);
+	len += scnprintf(buf + len, buf_len - len,
 			 "=================================================\n");
 
 	stats_req->buf_len = len;
--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.h
@@ -12,6 +12,10 @@
 #define HTT_STATS_MAGIC_VALUE   0xF0F0F0F0
 #define HTT_TLV_HDR_LEN                4
 
+#define HTT_NUM_AC_WMM	               0x4
+#define HTT_MAX_NUM_SBT_INTR	       4
+#define HTT_TX_CV_CORR_MAX_NUM_COLUMNS 8
+
 enum htt_tlv_tag_t {
 	HTT_STATS_TX_PDEV_CMN_TAG                           = 0,
 	HTT_STATS_TX_PDEV_UNDERRUN_TAG                      = 1,
@@ -875,6 +879,7 @@ enum htt_tx_mumimo_grp_invalid_reason_co
 
 #define HTT_TX_PDEV_STATS_NUM_AC_MUMIMO_USER_STATS    4
 #define HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS    8
+#define HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS    8
 #define HTT_TX_PDEV_STATS_NUM_OFDMA_USER_STATS       74
 #define HTT_TX_PDEV_STATS_NUM_UL_MUMIMO_USER_STATS    8
 #define HTT_STATS_MAX_MUMIMO_GRP_SZ		      8
@@ -905,6 +910,11 @@ struct htt_tx_selfgen_cmn_stats_tlv {
 	u32 su_sw_rts_err;
 	u32 su_sw_rts_flushed;
 	u32 su_sw_rts_rcvd_cts_diff_bw;
+	u32 combined_ax_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 combined_ax_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 standalone_ax_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 standalone_ax_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 smart_basic_trig_sch_histogram[HTT_MAX_NUM_SBT_INTR];
 };
 
 struct htt_tx_selfgen_ac_stats_tlv {
@@ -948,12 +958,25 @@ struct htt_tx_selfgen_ax_stats_tlv {
 	u32 ax_bsr_trigger;
 	u32 ax_mu_bar_trigger;
 	u32 ax_mu_rts_trigger;
+	u32 ax_ulmumimo_trigger;
 	u32 ax_su_ndpa_queued;
 	u32 ax_su_ndp_queued;
 	u32 ax_mu_mimo_ndpa_queued;
 	u32 ax_mu_mimo_ndp_queued;
 	u32 ax_mu_mimo_brpoll_queued[HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS - 1];
 	u32 ax_ul_mumimo_trigger[HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS];
+	u32 combined_ax_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 combined_ax_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 standalone_ax_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 standalone_ax_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 manual_ax_su_ulofdma_basic_trigger[HTT_NUM_AC_WMM];
+	u32 manual_ax_su_ulofdma_basic_trigger_err[HTT_NUM_AC_WMM];
+	u32 manual_ax_mu_ulofdma_basic_trigger[HTT_NUM_AC_WMM];
+	u32 manual_ax_mu_ulofdma_basic_trigger_err[HTT_NUM_AC_WMM];
+	u32 ax_basic_trigger_per_ac[HTT_NUM_AC_WMM];
+	u32 ax_basic_trigger_errors_per_ac[HTT_NUM_AC_WMM];
+	u32 ax_mu_bar_trigger_per_ac[HTT_NUM_AC_WMM];
+	u32 ax_mu_bar_trigger_errors_per_ac[HTT_NUM_AC_WMM];
 };
 
 struct htt_tx_selfgen_ac_err_stats_tlv {
@@ -1005,6 +1028,9 @@ struct htt_tx_selfgen_ax_err_stats_tlv {
 	u32 ax_mu_mimo_ndp_flushed;
 	u32 ax_mu_mimo_brpoll_flushed[HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS - 1];
 	u32 ax_ul_mumimo_trigger_err[HTT_TX_PDEV_STATS_NUM_AX_MUMIMO_USER_STATS];
+	u32 ax_basic_trigger_partial_resp;
+	u32 ax_bsr_trigger_partial_resp;
+	u32 ax_mu_bar_trigger_partial_resp;
 };
 
 /* == TX MU STATS == */
@@ -1767,6 +1793,8 @@ struct htt_tx_pdev_rate_stats_tlv {
 	u32 ax_su_embedded_trigger_data_ppdu_err;
 	/** sta side trigger stats */
 	u32 trigger_type_11be[HTT_TX_PDEV_STATS_NUM_11BE_TRIGGER_TYPES];
+	u32 extra_eht_ltf;
+	u32 extra_eht_ltf_ofdma;
 };
 
 /* == PDEV RX RATE CTRL STATS == */
@@ -1937,6 +1965,7 @@ struct htt_rx_pdev_rate_stats_tlv {
 	u32 rx_ulmumimo_mpdu_fail[HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER];
 	u32 rx_ulofdma_non_data_nusers[HTT_RX_PDEV_MAX_OFDMA_NUM_USER];
 	u32 rx_ulofdma_data_nusers[HTT_RX_PDEV_MAX_OFDMA_NUM_USER];
+	u32 rx_mcs_ext[HTT_RX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS];
 };
 
 struct htt_rx_pdev_rate_ext_stats_tlv {
@@ -2301,6 +2330,9 @@ enum htt_txbf_sound_steer_modes {
 enum htt_stats_sounding_tx_mode {
 	HTT_TX_AC_SOUNDING_MODE                      = 0,
 	HTT_TX_AX_SOUNDING_MODE                      = 1,
+	HTT_TX_BE_SOUNDING_MODE                      = 2,
+	HTT_TX_CMN_SOUNDING_MODE                     = 3,
+	HTT_TX_CV_CORR_MODE                          = 4,
 };
 
 struct htt_tx_sounding_stats_tlv {
@@ -2320,6 +2352,68 @@ struct htt_tx_sounding_stats_tlv {
 	 * ... where max_bw == 4 for 160mhz
 	 */
 	u32 sounding[HTT_TX_NUM_OF_SOUNDING_STATS_WORDS];
+	u32 cv_nc_mismatch_err;
+	u32 cv_fcs_err;
+	u32 cv_frag_idx_mismatch;
+	u32 cv_invalid_peer_id;
+	u32 cv_no_txbf_setup;
+	u32 cv_expiry_in_update;
+	u32 cv_pkt_bw_exceed;
+	u32 cv_dma_not_done_err;
+	u32 cv_update_failed;
+	u32 cv_total_query;
+	u32 cv_total_pattern_query;
+	u32 cv_total_bw_query;
+	u32 cv_invalid_bw_coding;
+	u32 cv_forced_sounding;
+	u32 cv_standalone_sounding;
+	u32 cv_nc_mismatch;
+	u32 cv_fb_type_mismatch;
+	u32 cv_ofdma_bw_mismatch;
+	u32 cv_bw_mismatch;
+	u32 cv_pattern_mismatch;
+	u32 cv_preamble_mismatch;
+	u32 cv_nr_mismatch;
+	u32 cv_in_use_cnt_exceeded;
+	u32 cv_found;
+	u32 cv_not_found;
+	u32 sounding_320[HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS];
+	u32 cbf_320[HTT_TXBF_MAX_NUM_OF_MODES];
+	u32 cv_ntbr_sounding;
+	u32 cv_found_upload_in_progress;
+	u32 cv_expired_during_query;
+	u32 cv_dma_timeout_error;
+	u32 cv_buf_ibf_uploads;
+	u32 cv_buf_ebf_uploads;
+	u32 cv_buf_received;
+	u32 cv_buf_fed_back;
+	u32 cv_total_query_ibf;
+	u32 cv_found_ibf;
+	u32 cv_not_found_ibf;
+	u32 cv_expired_during_query_ibf;
+	u32 adaptive_snd_total_query;
+	u32 adaptive_snd_total_mcs_drop[HTT_TX_PDEV_STATS_NUM_MCS_COUNTERS +
+					HTT_TX_PDEV_STATS_NUM_EXTRA_MCS_COUNTERS];
+	u32 adaptive_snd_kicked_in;
+	u32 adaptive_snd_back_to_default;
+	u32 cv_corr_trigger_online_mode;
+	u32 cv_corr_trigger_offline_mode;
+	u32 cv_corr_trigger_hybrid_mode;
+	u32 cv_corr_trigger_computation_level_0;
+	u32 cv_corr_trigger_computation_level_1;
+	u32 cv_corr_trigger_computation_level_2;
+	u32 cv_corr_trigger_num_users[HTT_TX_CV_CORR_MAX_NUM_COLUMNS];
+	u32 cv_corr_trigger_num_streams[HTT_TX_CV_CORR_MAX_NUM_COLUMNS];
+	u32 cv_corr_upload_total_buf_received;
+	u32 cv_corr_upload_total_buf_fed_back;
+	u32 cv_corr_upload_total_processing_failed;
+	u32 cv_corr_upload_failed_total_users_zero;
+	u32 cv_corr_upload_failed_total_users_exceeded;
+	u32 cv_corr_upload_failed_peer_not_found;
+	u32 cv_corr_upload_user_nss_exceeded;
+	u32 cv_corr_upload_invalid_lookup_index;
+	u32 cv_corr_upload_total_num_users[HTT_TX_CV_CORR_MAX_NUM_COLUMNS];
+	u32 cv_corr_upload_total_num_streams[HTT_TX_CV_CORR_MAX_NUM_COLUMNS];
 };
 
 struct htt_pdev_obss_pd_stats_tlv {
@@ -2420,6 +2514,18 @@ struct htt_tx_selfgen_be_stats_tlv {
 	u32 be_mu_mimo_ndp_queued;
 	u32 be_mu_mimo_brpoll_queued[HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS - 1];
 	u32 be_ul_mumimo_trigger[HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS];
+	u32 combined_be_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 combined_be_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 standalone_be_bsr_trigger_tried[HTT_NUM_AC_WMM];
+	u32 standalone_be_bsr_trigger_err[HTT_NUM_AC_WMM];
+	u32 manual_be_su_ulofdma_basic_trigger[HTT_NUM_AC_WMM];
+	u32 manual_be_su_ulofdma_basic_trigger_err[HTT_NUM_AC_WMM];
+	u32 manual_be_mu_ulofdma_basic_trigger[HTT_NUM_AC_WMM];
+	u32 manual_be_mu_ulofdma_basic_trigger_err[HTT_NUM_AC_WMM];
+	u32 be_basic_trigger_per_ac[HTT_NUM_AC_WMM];
+	u32 be_basic_trigger_errors_per_ac[HTT_NUM_AC_WMM];
+	u32 be_mu_bar_trigger_per_ac[HTT_NUM_AC_WMM];
+	u32 be_mu_bar_trigger_errors_per_ac[HTT_NUM_AC_WMM];
 };
 
 struct htt_tx_selfgen_be_err_stats_tlv {
@@ -2440,6 +2546,11 @@ struct htt_tx_selfgen_be_err_stats_tlv {
 	u32 be_mu_mimo_ndp_flushed;
 	u32 be_mu_mimo_brpoll_flushed[HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS - 1];
 	u32 be_ul_mumimo_trigger_err[HTT_TX_PDEV_STATS_NUM_BE_MUMIMO_USER_STATS];
+	u32 be_basic_trigger_partial_resp;
+	u32 be_bsr_trigger_partial_resp;
+	u32 be_mu_bar_trigger_partial_resp;
+	u32 be_mu_rts_trigger_blocked;
+	u32 be_bsr_trigger_blocked;
 };
 
 struct htt_tx_selfgen_be_sched_status_stats_tlv {
@@ -2542,6 +2653,18 @@ struct htt_rx_pdev_be_ul_trigger_stats_t
 	 * Trig power headroom for STA AID in same idx - UNIT(dB)
 	 */
 	u32 be_uplink_sta_power_headroom[HTT_RX_UL_MAX_UPLINK_RSSI_TRACK];
+
+	/**
+	 * Number of EHT UL OFDMA per-user responses containing only a QoS null in
+	 * response to basic trigger. Typically a data response is expected.
+	 */
+	u32 be_ul_ofdma_basic_trigger_rx_qos_null_only;
+
+	/* UL MLO Queue Depth Sharing Stats */
+	u32 ul_mlo_send_qdepth_params_count;
+	u32 ul_mlo_proc_qdepth_params_count;
+	u32 ul_mlo_proc_accepted_qdepth_params_count;
+	u32 ul_mlo_proc_discarded_qdepth_params_count;
 };
 
 struct htt_rx_pdev_be_ul_mimo_user_stats_tlv {
@@ -2596,6 +2719,13 @@ struct htt_rx_pdev_ul_mumimo_trig_be_sta
 	s8 be_rx_ul_mumimo_fd_rssi[HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER][HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS];
 	/* Average pilot EVM measued for RX UL TB PPDU */
 	s8 be_rx_ulmumimo_pilot_evm_dB_mean[HTT_RX_PDEV_MAX_ULMUMIMO_NUM_USER][HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS];
+	/** Number of times UL MUMIMO TB PPDUs received in a punctured mode */
+	u32 rx_ul_mumimo_punctured_mode[HTT_RX_PDEV_STATS_NUM_PUNCTURED_MODE_COUNTERS];
+	/**
+	 * Number of EHT UL MU-MIMO per-user responses containing only a QoS null
+	 * in response to basic trigger. Typically a data response is expected.
+	 */
+	u32 be_ul_mumimo_basic_trigger_rx_qos_null_only;
 };
 
 #ifdef CPTCFG_ATH12K_DEBUGFS
@@ -2647,6 +2777,7 @@ struct htt_rx_pdev_ul_mumimo_trig_stats_
 					 [HTT_RX_PDEV_STATS_ULMUMIMO_NUM_SPATIAL_STREAMS];
 	u32 reduced_ul_mumimo_rx_bw[HTT_RX_PDEV_STATS_NUM_REDUCED_CHAN_TYPES]
 				   [HTT_RX_PDEV_STATS_NUM_BW_COUNTERS];
+	u32 ul_mumimo_basic_trigger_rx_qos_null_only;
 };
 
 struct htt_rx_pdev_ul_mimo_user_stats_tlv {
@@ -2704,6 +2835,11 @@ struct htt_latency_prof_stats_tlv {
 	u32 avg;
 	u32 hist_intvl;
 	u32 hist[HTT_LATENCY_PROFILE_MAX_HIST];
+	u32 page_fault_max;
+	u32 page_fault_total;
+	u32 ignored_latency_count;
+	u32 interrupts_max;
+	u32 interrupts_hist[HTT_INTERRUPTS_LATENCY_PROFILE_MAX_HIST];
 };
 
 struct htt_latency_prof_ctx_tlv {
@@ -2963,6 +3099,11 @@ struct htt_rx_pdev_be_ul_ofdma_user_stat
 #define HTT_STATS_MAX_CHAINS 8
 #define HTT_MAX_CH_PWR_INFO_SIZE 16
 #define HTT_PUNCTURE_STATS_MAX_SUBBAND_COUNT 32
+#define HTT_MAX_RX_PKT_CNT_EXT 4
+#define HTT_MAX_RX_PKT_CRC_PASS_CNT_EXT 4
+#define HTT_MAX_RX_PKT_MU_CNT 14
+#define HTT_MAX_TX_PKT_CNT 10
+#define HTT_MAX_PHY_TX_ABORT_CNT 10
 
 #define HTT_PDEV_PUNCTURE_STATS_MAC_ID_M 0x000000ff
 #define HTT_PDEV_PUNCTURE_STATS_MAC_ID_S 0
@@ -2976,6 +3117,66 @@ struct htt_rx_pdev_be_ul_ofdma_user_stat
 		((_var) |= ((_val) << HTT_PDEV_PUNCTURE_STATS_MAC_ID_S)); \
 	} while (0)
 
+#define HTT_PHY_TPC_STATS_CTL_REGION_GRP_M    0x000000ff
+#define HTT_PHY_TPC_STATS_CTL_REGION_GRP_S    0
+
+#define HTT_PHY_TPC_STATS_CTL_REGION_GRP_GET(_var) \
+	(((_var) & HTT_PHY_TPC_STATS_CTL_REGION_GRP_M) >> \
+	  HTT_PHY_TPC_STATS_CTL_REGION_GRP_S)
+#define HTT_STATS_PHY_TPC_STATS_CTL_REGION_GRP_GET(_var) \
+	HTT_PHY_TPC_STATS_CTL_REGION_GRP_GET(_var)
+#define HTT_PHY_TPC_STATS_CTL_REGION_GRP_SET(_var, _val) \
+	do { \
+		HTT_CHECK_SET_VAL(HTT_PHY_TPC_STATS_CTL_REGION_GRP, _val); \
+		((_var) &= ~(HTT_PHY_TPC_STATS_CTL_REGION_GRP_M)); \
+		((_var) |= ((_val) << HTT_PHY_TPC_STATS_CTL_REGION_GRP_S)); \
+	} while (0)
+
+#define HTT_PHY_TPC_STATS_SUB_BAND_INDEX_M    0x0000ff00
+#define HTT_PHY_TPC_STATS_SUB_BAND_INDEX_S    8
+
+#define HTT_PHY_TPC_STATS_SUB_BAND_INDEX_GET(_var) \
+	(((_var) & HTT_PHY_TPC_STATS_SUB_BAND_INDEX_M) >> \
+	  HTT_PHY_TPC_STATS_SUB_BAND_INDEX_S)
+#define HTT_STATS_PHY_TPC_STATS_SUB_BAND_INDEX_GET(_var) \
+	HTT_PHY_TPC_STATS_SUB_BAND_INDEX_GET(_var)
+#define HTT_PHY_TPC_STATS_SUB_BAND_INDEX_SET(_var, _val) \
+	do { \
+		HTT_CHECK_SET_VAL(HTT_PHY_TPC_STATS_SUB_BAND_INDEX, _val); \
+		((_var) &= ~(HTT_PHY_TPC_STATS_SUB_BAND_INDEX_M)); \
+		((_var) |= ((_val) << HTT_PHY_TPC_STATS_SUB_BAND_INDEX_S)); \
+	} while (0)
+
+#define HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_M    0x00ff0000
+#define HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_S    16
+
+#define HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_GET(_var) \
+	(((_var) & HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_M) >> \
+	  HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_S)
+#define HTT_STATS_PHY_TPC_STATS_ARRAY_GAIN_CAP_EXT2_ENABLED_GET(_var) \
+	HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_GET(_var)
+#define HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_SET(_var, _val) \
+	do { \
+		HTT_CHECK_SET_VAL(HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED, _val); \
+		((_var) &= ~(HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_M)); \
+		((_var) |= ((_val) << HTT_PHY_TPC_STATS_AG_CAP_EXT2_ENABLED_S)); \
+	} while (0)
+
+#define HTT_PHY_TPC_STATS_CTL_FLAG_M    0xff000000
+#define HTT_PHY_TPC_STATS_CTL_FLAG_S    24
+
+#define HTT_PHY_TPC_STATS_CTL_FLAG_GET(_var) \
+	(((_var) & HTT_PHY_TPC_STATS_CTL_FLAG_M) >> \
+	  HTT_PHY_TPC_STATS_CTL_FLAG_S)
+#define HTT_STATS_PHY_TPC_STATS_CTL_FLAG_GET(_var) \
+	HTT_PHY_TPC_STATS_CTL_FLAG_GET(_var)
+#define HTT_PHY_TPC_STATS_CTL_FLAG_SET(_var, _val) \
+	do { \
+		HTT_CHECK_SET_VAL(HTT_PHY_TPC_STATS_CTL_FLAG, _val); \
+		((_var) &= ~(HTT_PHY_TPC_STATS_CTL_FLAG_M)); \
+		((_var) |= ((_val) << HTT_PHY_TPC_STATS_CTL_FLAG_S)); \
+	} while (0)
+
 struct htt_phy_tpc_stats_tlv {
 	u32 pdev_id;
 	u32 tx_power_scale;
@@ -2992,6 +3193,16 @@ struct htt_phy_tpc_stats_tlv {
 	u32 power_type_6ghz;
 	u32 sub_band_cfreq[HTT_MAX_CH_PWR_INFO_SIZE];
 	u32 sub_band_txpower[HTT_MAX_CH_PWR_INFO_SIZE];
+	s32 array_gain_cap[HTT_STATS_MAX_CHAINS * ((HTT_STATS_MAX_CHAINS / 2) + 1)];
+	union {
+		struct {
+			u32 ctl_region_grp:8,
+			    sub_band_index:8,
+			    array_gain_cap_ext2_enabled:8,
+			    ctl_flag:8;
+		};
+		u32 ctl_args;
+	};
 };
 
 struct htt_pdev_puncture_stats_tlv {
@@ -3645,6 +3856,11 @@ struct htt_phy_counters_tlv {
 	u32 rx_pkt_crc_pass_cnt[HTT_MAX_RX_PKT_CRC_PASS_CNT];
 	u32 per_blk_err_cnt[HTT_MAX_PER_BLK_ERR_CNT];
 	u32 rx_ota_err_cnt[HTT_MAX_RX_OTA_ERR_CNT];
+	u32 rx_pkt_cnt_ext[HTT_MAX_RX_PKT_CNT_EXT];
+	u32 rx_pkt_crc_pass_cnt_ext[HTT_MAX_RX_PKT_CRC_PASS_CNT_EXT];
+	u32 rx_pkt_mu_cnt[HTT_MAX_RX_PKT_MU_CNT];
+	u32 tx_pkt_cnt[HTT_MAX_TX_PKT_CNT];
+	u32 phy_tx_abort_cnt[HTT_MAX_PHY_TX_ABORT_CNT];
 };
 
 struct htt_phy_stats_tlv {
@@ -3672,6 +3888,7 @@ struct htt_phy_reset_counters_tlv {
 	u32 phy_off_through_vreg_cnt;
 	u32 force_calibration_cnt;
 	u32 rf_mode_switch_phy_off_cnt;
+	u32 temperature_recal_cnt;
 };
 
 struct htt_phy_reset_stats_tlv {
@@ -3704,6 +3921,14 @@ struct htt_phy_reset_stats_tlv {
 	u32 dyn_cca_status;
 	u32 rxdesense_thresh_hw;
 	u32 rxdesense_thresh_sw;
+	u32 phy_bw_code;
+	u32 phy_rate_mode;
+	u32 phy_band_code;
+	u32 phy_vreg_base;
+	u32 phy_vreg_base_ext;
+	u32 cur_table_index;
+	u32 whal_config_flag;
+	u32 nfcal_iteration_counts[3];
 };
 
 struct htt_peer_ctrl_path_txrx_stats_tlv {
@@ -3981,6 +4206,9 @@ struct htt_tx_selfgen_ax_sched_status_st
 enum htt_stats_rc_mode {
 	HTT_STATS_RC_MODE_DLSU	   = 0,
 	HTT_STATS_RC_MODE_DLMUMIMO = 1,
+	HTT_STATS_RC_MODE_DLOFDMA  = 2,
+	HTT_STATS_RC_MODE_ULMUMIMO = 3,
+	HTT_STATS_RC_MODE_ULOFDMA  = 4,
 };
 
 struct htt_tx_rate_stats_t {
@@ -3990,6 +4218,25 @@ struct htt_tx_rate_stats_t {
 	u32 mpdus_failed;
 };
 
+enum HTT_RC_MODE {
+	HTT_RC_MODE_SU_OL,
+	HTT_RC_MODE_SU_BF,
+	HTT_RC_MODE_MU1_INTF,
+	HTT_RC_MODE_MU2_INTF,
+	HTT_RC_MODE_MU3_INTF,
+	HTT_RC_MODE_MU4_INTF,
+	HTT_RC_MODE_MU5_INTF,
+	HTT_RC_MODE_MU6_INTF,
+	HTT_RC_MODE_MU7_INTF,
+	HTT_RC_MODE_2D_COUNT,
+};
+
+enum htt_stats_ru_type {
+	HTT_STATS_RU_TYPE_INVALID,
+	HTT_STATS_RU_TYPE_SINGLE_RU_ONLY,
+	HTT_STATS_RU_TYPE_SINGLE_AND_MULTI_RU,
+};
+
 struct htt_tx_per_rate_stats_tlv {
 	u32 rc_mode;
 	u32 last_probed_mcs;
@@ -4002,6 +4249,11 @@ struct htt_tx_per_rate_stats_tlv {
 
 	/** 320MHz extension for PER */
 	struct htt_tx_rate_stats_t per_bw320;
+
+	u32 probe_cnt_per_rcmode[HTT_RC_MODE_2D_COUNT];
+	u32 ru_type;
+
+	struct htt_tx_rate_stats_t per_ru[HTT_TX_PDEV_STATS_NUM_BE_RU_SIZE_COUNTERS];
 };
 
 struct htt_tx_pdev_dl_mu_ofdma_sch_stats_tlv {
@@ -4089,8 +4341,6 @@ struct htt_tx_pdev_ul_mu_mimo_sch_stats_
 /* 0 - Half, 1 - Quarter */
 #define HTT_STA_UL_OFDMA_NUM_REDUCED_CHAN_TYPES 2
 
-#define HTT_NUM_AC_WMM	0x4
-
 enum HTT_STA_UL_OFDMA_RX_TRIG_TYPE {
 	HTT_ULTRIG_QBOOST_TRIGGER = 0,
 	HTT_ULTRIG_PSPOLL_TRIGGER,
@@ -4220,6 +4470,9 @@ struct htt_peer_ax_ofdma_stats_tlv {
 	u32 last_updated_ul_qdepth[HTT_NUM_AC_WMM];
 };
 
+#define HTT_STATS_SCHED_OFDMA_TXBF_INELIGIBILITY_MAX 8
+#define HTT_MAX_NUM_CHAN_ACC_LAT_INTR 9
+
 struct htt_pdev_sched_algo_ofdma_stats_tlv {
 	/**
 	 * BIT [ 7 :  0]   :- mac_id
@@ -4246,6 +4499,8 @@ struct htt_pdev_sched_algo_ofdma_stats_t
 	u32 dlofdma_disabled_su_only_eligible[HTT_NUM_AC_WMM];
 	u32 dlofdma_disabled_consec_no_mpdus_tried[HTT_NUM_AC_WMM];
 	u32 dlofdma_disabled_consec_no_mpdus_success[HTT_NUM_AC_WMM];
+	u32 txbf_ofdma_ineligibility_stat[HTT_STATS_SCHED_OFDMA_TXBF_INELIGIBILITY_MAX];
+	u32 avg_chan_acc_lat_hist[HTT_MAX_NUM_CHAN_ACC_LAT_INTR];
 };
 
 struct htt_umac_ssr_stats_tlv {
